diff --git a/include/vkd3d.h b/include/vkd3d.h
index bc77c8c..69b1d68 100644
--- a/include/vkd3d.h
+++ b/include/vkd3d.h
@@ -95,7 +95,7 @@ extern "C" {
 #define VKD3D_CONFIG_FLAG_FORCE_DEDICATED_IMAGE_ALLOCATION (1ull << 35)
 #define VKD3D_CONFIG_FLAG_BREADCRUMBS_TRACE (1ull << 36)
 #define VKD3D_CONFIG_FLAG_DISABLE_SIMULTANEOUS_UAV_COMPRESSION (1ull << 37)
-#define VKD3D_CONFIG_FLAG_HUGE_NV_DGC_BUFFERS (1ull << 38)
+#define VKD3D_CONFIG_FLAG_REQUIRES_COMPUTE_INDIRECT_TEMPLATES (1ull << 38)
 #define VKD3D_CONFIG_FLAG_SKIP_DRIVER_WORKAROUNDS (1ull << 39)
 #define VKD3D_CONFIG_FLAG_RETAIN_PSOS (1ull << 40)
 #define VKD3D_CONFIG_FLAG_ENABLE_EXPERIMENTAL_FEATURES (1ull << 41)
@@ -103,7 +103,7 @@ extern "C" {
 #define VKD3D_CONFIG_FLAG_DISABLE_UAV_COMPRESSION (1ull << 43)
 #define VKD3D_CONFIG_FLAG_DISABLE_DEPTH_COMPRESSION (1ull << 44)
 #define VKD3D_CONFIG_FLAG_DISABLE_COLOR_COMPRESSION (1ull << 45)
-/* Bit 46 is vacant */
+#define VKD3D_CONFIG_FLAG_DISABLE_DGCC (1ull << 46)
 #define VKD3D_CONFIG_FLAG_MEMORY_ALLOCATOR_SKIP_IMAGE_HEAP_CLEAR (1ull << 47)
 #define VKD3D_CONFIG_FLAG_DRIVER_VERSION_SENSITIVE_SHADERS (1ull << 48)
 #define VKD3D_CONFIG_FLAG_SMALL_VRAM_REBAR (1ull << 49)
diff --git a/libs/vkd3d/command.c b/libs/vkd3d/command.c
index 84976e6..9bd4ed6 100644
--- a/libs/vkd3d/command.c
+++ b/libs/vkd3d/command.c
@@ -6373,7 +6373,8 @@ static void vk_access_and_stage_flags_from_d3d12_resource_state(const struct d3d
                 *access |= VK_ACCESS_2_INDIRECT_COMMAND_READ_BIT;
 
                 /* We might use preprocessing. */
-                if (list->device->device_info.device_generated_commands_features.deviceGeneratedCommands)
+                if (list->device->device_info.device_generated_commands_features_nv.deviceGeneratedCommands ||
+                    list->device->device_info.device_generated_commands_features_ext.deviceGeneratedCommands)
                 {
                     *stages |= VK_PIPELINE_STAGE_2_COMMAND_PREPROCESS_BIT_EXT;
                     *access |= VK_ACCESS_2_COMMAND_PREPROCESS_READ_BIT_EXT;
@@ -8137,6 +8138,52 @@ static void d3d12_command_list_update_descriptors(struct d3d12_command_list *lis
     }
 }
 
+static void d3d12_command_list_update_descriptors_post_indirect_buffer(struct d3d12_command_list *list)
+{
+    /* Pretend for a moment that the post indirect buffer is the main command buffer.
+     * Set all dirty bits so we force-flush state to a different command buffer.
+     * Not the most elegant solution, but avoids us having to reimplement everything
+     * just to plumb thorugh a different set of dirty masks, etc. */
+    struct vkd3d_pipeline_bindings *bindings = d3d12_command_list_get_bindings(list, list->active_pipeline_type);
+    const struct d3d12_root_signature *rs = bindings->root_signature;
+    uint32_t old_root_descriptor_dirty_mask;
+    uint32_t old_descriptor_heap_dirty_mask;
+    uint32_t old_root_constant_dirty_mask;
+    VkCommandBuffer old_cmd_buffer;
+    bool old_heap_dirty = false;
+    uint32_t old_dirty_flags;
+
+    if (!rs)
+        return;
+
+    old_root_descriptor_dirty_mask = bindings->root_descriptor_dirty_mask;
+    old_descriptor_heap_dirty_mask = bindings->descriptor_heap_dirty_mask;
+    old_root_constant_dirty_mask = bindings->root_constant_dirty_mask;
+    old_dirty_flags = bindings->dirty_flags;
+    old_cmd_buffer = list->cmd.vk_command_buffer;
+    /* This is bad, but the current NV implementation does not actually
+     * do anything bad when rebinding descriptor buffers, so just roll with it.
+     * Can be fixed if necessary. */
+    if (d3d12_device_uses_descriptor_buffers(list->device))
+        old_heap_dirty = list->descriptor_heap.buffers.heap_dirty;
+
+    /* Override state. */
+    list->cmd.vk_command_buffer = list->cmd.vk_post_indirect_barrier_commands;
+    if (d3d12_device_uses_descriptor_buffers(list->device))
+        list->descriptor_heap.buffers.heap_dirty = true;
+    d3d12_command_list_invalidate_root_parameters(list, bindings, true, NULL);
+    d3d12_command_list_update_descriptors(list);
+
+    /* Restore state. */
+    bindings->root_descriptor_dirty_mask = old_root_descriptor_dirty_mask;
+    bindings->descriptor_heap_dirty_mask = old_descriptor_heap_dirty_mask;
+    bindings->root_constant_dirty_mask = old_root_constant_dirty_mask;
+    bindings->dirty_flags = old_dirty_flags;
+    list->cmd.vk_command_buffer = old_cmd_buffer;
+    if (d3d12_device_uses_descriptor_buffers(list->device))
+        list->descriptor_heap.buffers.heap_dirty = old_heap_dirty;
+}
+
 static void d3d12_command_list_check_pre_compute_barrier(
         struct d3d12_command_list *list, VkPipelineStageFlagBits2 vk_dst_stage);
 
@@ -8833,6 +8880,100 @@ static bool d3d12_command_list_emit_multi_dispatch_indirect_count(struct d3d12_c
     return true;
 }
 
+static bool d3d12_command_list_emit_multi_dispatch_indirect_count_state(struct d3d12_command_list *list,
+        struct d3d12_command_signature *signature,
+        VkDeviceAddress indirect_args,
+        uint32_t stride, uint32_t max_commands,
+        VkDeviceAddress count_arg,
+        struct vkd3d_scratch_allocation *dispatch_scratch,
+        struct vkd3d_scratch_allocation *ubo_scratch)
+{
+    const struct vkd3d_vk_device_procs *vk_procs = &list->device->vk_procs;
+    struct vkd3d_multi_dispatch_indirect_info pipeline_info;
+    struct vkd3d_multi_dispatch_indirect_state_args args;
+    struct vkd3d_scratch_allocation template_scratch;
+    VkCommandBuffer vk_patch_cmd_buffer;
+    VkMemoryBarrier2 vk_barrier;
+    VkDependencyInfo dep_info;
+
+    vkd3d_meta_get_multi_dispatch_indirect_state_pipeline(&list->device->meta_ops, &pipeline_info);
+
+    if (!d3d12_command_allocator_allocate_scratch_memory(list->allocator,
+            VKD3D_SCRATCH_POOL_KIND_UNIFORM_UPLOAD,
+            D3D12_MAX_ROOT_COST * sizeof(uint32_t) +
+                    sizeof(signature->state_template.compute.source_offsets),
+            sizeof(uint32_t), ~0u, &template_scratch))
+        return false;
+
+    if (!d3d12_command_allocator_allocate_scratch_memory(list->allocator,
+            VKD3D_SCRATCH_POOL_KIND_DEVICE_STORAGE,
+            sizeof(VkDispatchIndirectCommand) * max_commands,
+            sizeof(uint32_t), ~0u, dispatch_scratch))
+        return false;
+
+    if (!d3d12_command_allocator_allocate_scratch_memory(list->allocator,
+            VKD3D_SCRATCH_POOL_KIND_DEVICE_STORAGE,
+            (D3D12_MAX_ROOT_COST * sizeof(uint32_t)) * max_commands,
+            D3D12_CONSTANT_BUFFER_DATA_PLACEMENT_ALIGNMENT,
+            ~0u, ubo_scratch))
+        return false;
+
+    d3d12_command_list_fetch_root_parameter_data(list, &list->compute_bindings, template_scratch.host_ptr);
+    memcpy(void_ptr_offset(template_scratch.host_ptr, D3D12_MAX_ROOT_COST * sizeof(uint32_t)),
+            signature->state_template.compute.source_offsets,
+            sizeof(signature->state_template.compute.source_offsets));
+
+    args.indirect_va = indirect_args;
+    args.count_va = count_arg;
+    args.dispatch_va = dispatch_scratch->va;
+    args.root_parameters_va = ubo_scratch->va;
+    args.root_parameter_template_va = template_scratch.va;
+    args.stride_words = stride / sizeof(uint32_t);
+    args.dispatch_offset_words = signature->state_template.compute.dispatch_offset_words;
+
+    d3d12_command_allocator_allocate_init_post_indirect_command_buffer(list->allocator, list);
+    vk_patch_cmd_buffer = list->cmd.vk_post_indirect_barrier_commands;
+
+    if (vk_patch_cmd_buffer == list->cmd.vk_command_buffer)
+    {
+        d3d12_command_list_invalidate_current_pipeline(list, true);
+        d3d12_command_list_invalidate_root_parameters(list, &list->compute_bindings, true, &list->graphics_bindings);
+    }
+    else
+    {
+        list->cmd.indirect_meta->need_compute_to_indirect_barrier = true;
+        list->cmd.indirect_meta->need_compute_to_cbv_barrier = true;
+    }
+
+    VK_CALL(vkCmdBindPipeline(vk_patch_cmd_buffer, VK_PIPELINE_BIND_POINT_COMPUTE,
+            pipeline_info.vk_pipeline));
+    VK_CALL(vkCmdPushConstants(vk_patch_cmd_buffer,
+            pipeline_info.vk_pipeline_layout, VK_SHADER_STAGE_COMPUTE_BIT,
+            0, sizeof(args), &args));
+
+    VK_CALL(vkCmdDispatch(vk_patch_cmd_buffer, max_commands, 1, 1));
+
+    if (vk_patch_cmd_buffer == list->cmd.vk_command_buffer)
+    {
+        memset(&dep_info, 0, sizeof(dep_info));
+        dep_info.sType = VK_STRUCTURE_TYPE_DEPENDENCY_INFO;
+        dep_info.memoryBarrierCount = 1;
+        dep_info.pMemoryBarriers = &vk_barrier;
+
+        memset(&vk_barrier, 0, sizeof(vk_barrier));
+        vk_barrier.sType = VK_STRUCTURE_TYPE_MEMORY_BARRIER_2;
+        vk_barrier.srcStageMask = VK_PIPELINE_STAGE_2_COMPUTE_SHADER_BIT;
+        vk_barrier.srcAccessMask = VK_ACCESS_2_SHADER_WRITE_BIT;
+        vk_barrier.dstStageMask = VK_PIPELINE_STAGE_2_DRAW_INDIRECT_BIT | VK_PIPELINE_STAGE_2_COMPUTE_SHADER_BIT;
+        vk_barrier.dstAccessMask = VK_ACCESS_2_INDIRECT_COMMAND_READ_BIT | VK_ACCESS_2_UNIFORM_READ_BIT;
+
+        VK_CALL(vkCmdPipelineBarrier2(vk_patch_cmd_buffer, &dep_info));
+    }
+
+    VKD3D_BREADCRUMB_COMMAND(EXECUTE_INDIRECT_PATCH_STATE_COMPUTE);
+    return true;
+}
+
 static bool d3d12_command_list_emit_predicated_command(struct d3d12_command_list *list,
         enum vkd3d_predicate_command_type command_type, VkDeviceAddress indirect_args,
         const union vkd3d_predicate_command_direct_args *direct_args, struct vkd3d_scratch_allocation *scratch)
@@ -16268,6 +16409,18 @@ static void STDMETHODCALLTYPE d3d12_command_list_SetPredication(d3d12_command_li
             vk_barrier.dstStageMask = VK_PIPELINE_STAGE_2_CONDITIONAL_RENDERING_BIT_EXT;
             vk_barrier.dstAccessMask = VK_ACCESS_2_CONDITIONAL_RENDERING_READ_BIT_EXT;
             list->predication.enabled_on_command_buffer = true;
+
+            /* EXT_conditional_rendering does not play nice with NV_dgc yet
+             * (somewhat ambiguous what the spec means, but NV handles it),
+             * so we will need a fallback there where we read from the
+             * predicate VA. INDIRECT_ACCESS barriers on Mesa imply SCACHE/VCACHE anyway, so this does not really hurt us. */
+            if (!(vkd3d_config_flags & VKD3D_CONFIG_FLAG_SKIP_DRIVER_WORKAROUNDS) &&
+                    !list->device->device_info.device_generated_commands_features_ext.deviceGeneratedCommands &&
+                    list->device->device_info.vulkan_1_2_properties.driverID == VK_DRIVER_ID_MESA_RADV)
+            {
+                vk_barrier.dstStageMask |= VK_PIPELINE_STAGE_2_COMPUTE_SHADER_BIT;
+                vk_barrier.dstAccessMask |= VK_ACCESS_2_SHADER_READ_BIT;
+            }
         }
         else
         {
@@ -16548,13 +16701,114 @@ static HRESULT d3d12_command_signature_allocate_stream_memory_for_list(
         uint32_t max_command_count,
         struct vkd3d_scratch_allocation *allocation);
 
-static HRESULT d3d12_command_signature_allocate_preprocess_memory_for_list(
+static HRESULT d3d12_command_signature_allocate_preprocess_memory_for_list_ext(
+        struct d3d12_command_list *list,
+        struct d3d12_command_signature *signature, VkPipeline render_pipeline,
+        bool explicit_preprocess,
+        uint32_t max_command_count,
+        struct vkd3d_scratch_allocation *allocation, VkDeviceSize *size);
+
+static HRESULT d3d12_command_signature_allocate_preprocess_memory_for_list_nv(
         struct d3d12_command_list *list,
         struct d3d12_command_signature *signature, VkPipeline render_pipeline,
         bool explicit_preprocess,
         uint32_t max_command_count,
         struct vkd3d_scratch_allocation *allocation, VkDeviceSize *size);
 
+static void d3d12_command_list_execute_indirect_state_template_compute(
+        struct d3d12_command_list *list, struct d3d12_command_signature *signature,
+        uint32_t max_command_count,
+        struct d3d12_resource *arg_buffer, UINT64 arg_buffer_offset,
+        struct d3d12_resource *count_buffer, UINT64 count_buffer_offset)
+{
+    const struct vkd3d_vk_device_procs *vk_procs = &list->device->vk_procs;
+    VkDeviceAddress arg_va = arg_buffer->res.va + arg_buffer_offset;
+    struct vkd3d_scratch_allocation dispatch_scratch, ubo_scratch;
+    VkDeviceAddress count_va = 0;
+    VkWriteDescriptorSet write;
+    VkDescriptorBufferInfo buf;
+    VkPipelineLayout vk_layout;
+    uint32_t write_set;
+    unsigned int i;
+
+    d3d12_command_list_end_current_render_pass(list, false);
+    d3d12_command_list_end_transfer_batch(list);
+
+    /* If this command breaks suspend, need to refresh it now. */
+    d3d12_command_list_update_conditional_rendering_state(list, false);
+
+    if (count_buffer)
+        count_va = count_buffer->res.va + count_buffer_offset;
+
+    if (!d3d12_command_list_emit_multi_dispatch_indirect_count_state(list,
+            signature,
+            arg_va, signature->desc.ByteStride, max_command_count,
+            count_va, &dispatch_scratch, &ubo_scratch))
+        return;
+
+    if (!d3d12_command_list_update_compute_state(list))
+    {
+        WARN("Failed to update compute state, ignoring dispatch.\n");
+        return;
+    }
+
+    vk_write_descriptor_set_from_scratch_push_ubo(&write, &buf, &ubo_scratch,
+            D3D12_MAX_ROOT_COST * sizeof(uint32_t),
+            list->compute_bindings.root_signature->push_constant_ubo_binding.binding);
+
+    vk_layout = list->compute_bindings.root_signature->compute.vk_pipeline_layout;
+    write_set = list->compute_bindings.root_signature->push_constant_ubo_binding.set;
+
+    /* Run indirect dispatches back to back with one push UBO per dispatch which lets us
+     * update root parameters per command. */
+    for (i = 0; i < max_command_count; i++)
+    {
+        VK_CALL(vkCmdPushDescriptorSetKHR(list->cmd.vk_command_buffer, VK_PIPELINE_BIND_POINT_COMPUTE,
+                vk_layout, write_set, 1, &write));
+        VK_CALL(vkCmdDispatchIndirect(list->cmd.vk_command_buffer, dispatch_scratch.buffer, dispatch_scratch.offset));
+
+        VKD3D_BREADCRUMB_AUX32(i);
+        VKD3D_BREADCRUMB_COMMAND(EXECUTE_INDIRECT_UNROLL_COMPUTE);
+
+        dispatch_scratch.offset += sizeof(VkDispatchIndirectCommand);
+        buf.offset += D3D12_MAX_ROOT_COST * sizeof(uint32_t);
+    }
+
+    /* Need to clear state to zero if it was part of a command signature. */
+    for (i = 0; i < signature->desc.NumArgumentDescs; i++)
+    {
+        const D3D12_INDIRECT_ARGUMENT_DESC *arg = &signature->desc.pArgumentDescs[i];
+        switch (arg->Type)
+        {
+            case D3D12_INDIRECT_ARGUMENT_TYPE_CONSTANT_BUFFER_VIEW:
+            case D3D12_INDIRECT_ARGUMENT_TYPE_SHADER_RESOURCE_VIEW:
+            case D3D12_INDIRECT_ARGUMENT_TYPE_UNORDERED_ACCESS_VIEW:
+            {
+                uint32_t index = arg->ConstantBufferView.RootParameterIndex;
+                d3d12_command_list_set_root_descriptor(list,
+                        &list->compute_bindings, index, 0);
+                break;
+            }
+
+            case D3D12_INDIRECT_ARGUMENT_TYPE_CONSTANT:
+            {
+                uint32_t zeroes[D3D12_MAX_ROOT_COST];
+                memset(zeroes, 0, sizeof(uint32_t) * arg->Constant.Num32BitValuesToSet);
+                d3d12_command_list_set_root_constants(list,
+                        &list->compute_bindings, arg->Constant.RootParameterIndex,
+                        arg->Constant.DestOffsetIn32BitValues,
+                        arg->Constant.Num32BitValuesToSet, zeroes);
+                break;
+            }
+
+            default:
+                break;
+        }
+    }
+
+    /* No need to implicitly invalidate anything here, since we used the normal APIs. */
+}
+
 enum vkd3d_dgc_mode
 {
     /* When ExecuteIndirect is being called directly by application. */
@@ -16578,12 +16832,16 @@ static void d3d12_command_list_execute_indirect_state_template_dgc(
     struct vkd3d_scratch_allocation predication_allocation;
     struct vkd3d_scratch_allocation preprocess_allocation;
     struct vkd3d_scratch_allocation stream_allocation;
+    uint32_t minIndirectCommandsBufferOffsetAlignment;
     struct vkd3d_scratch_allocation count_allocation;
     VkGeneratedCommandsPipelineInfoEXT pipeline_info;
+    uint32_t minSequencesCountBufferOffsetAlignment;
     bool old_predication_enabled_on_command_buffer;
     struct vkd3d_execute_indirect_args patch_args;
     VkGeneratedCommandsInfoEXT generated_ext;
+    VkGeneratedCommandsInfoNV generated_nv;
     VkCommandBuffer vk_patch_cmd_buffer;
+    VkIndirectCommandsStreamNV stream;
     bool require_custom_predication;
     VkPipeline current_pipeline;
     VkDependencyInfo dep_info;
@@ -16592,10 +16850,12 @@ static void d3d12_command_list_execute_indirect_state_template_dgc(
     bool explicit_preprocess;
     bool require_ibo_update;
     bool require_patch;
+    bool use_ext_dgc;
     const char *tag;
     unsigned int i;
     HRESULT hr;
 
+    use_ext_dgc = list->device->device_info.device_generated_commands_features_ext.deviceGeneratedCommands;
     require_custom_predication = false;
     restart_predication = false;
     explicit_preprocess = false;
@@ -16605,7 +16865,13 @@ static void d3d12_command_list_execute_indirect_state_template_dgc(
     {
         bool fallback_predication = false;
 
-        if (!list->cmd.suspend_resume.block_resume && list->device->workarounds.tiler_suspend_resume)
+        if (!(vkd3d_config_flags & VKD3D_CONFIG_FLAG_SKIP_DRIVER_WORKAROUNDS) &&
+                !list->device->device_info.device_generated_commands_features_ext.deviceGeneratedCommands &&
+                list->device->device_info.vulkan_1_2_properties.driverID == VK_DRIVER_ID_MESA_RADV)
+        {
+            fallback_predication = true;
+        }
+        else if (!list->cmd.suspend_resume.block_resume && list->device->workarounds.tiler_suspend_resume)
         {
             /* We have not yet resumed. If we set the necessary state on this command buffer, we will break resume.
              * Instead, use fallback path which avoids this problem.
@@ -16661,7 +16927,8 @@ static void d3d12_command_list_execute_indirect_state_template_dgc(
         explicit_preprocess = false;
     }
     else if (!require_custom_predication && list->dgc_batch.draws_count == 0 &&
-            signature->state_template.dgc.layout_preprocess)
+            (signature->state_template.dgc.layout_preprocess_nv ||
+             signature->state_template.dgc.layout_preprocess_ext))
     {
         /* If driver can take advantage of preprocess, we can consider preprocessing explicitly if we can hoist it.
          * If we had indirect barriers earlier in the frame, now might be a good time to split. */
@@ -16672,7 +16939,7 @@ static void d3d12_command_list_execute_indirect_state_template_dgc(
 
     /* Don't try to batch if we have predication since it's just needless complexity.
      * Don't do this on tilers since we'll be introducing more technically unnecessary render pass flushes. */
-    if (!list->predication.va && dgc_mode == VKD3D_DGC_MODE_APPLICATION_CALL &&
+    if (!list->predication.va && dgc_mode == VKD3D_DGC_MODE_APPLICATION_CALL && use_ext_dgc &&
         !list->device->workarounds.tiler_suspend_resume &&
         !explicit_preprocess && list->state && list->state->pipeline_type != VKD3D_PIPELINE_TYPE_COMPUTE)
     {
@@ -16743,7 +17010,27 @@ static void d3d12_command_list_execute_indirect_state_template_dgc(
         }
     }
 
+    /* - Stride can mismatch, i.e. we need internal alignment of arguments.
+     * - Min required alignment on the indirect buffer itself might be too strict.
+     * - Min required alignment on count buffer might be too strict.
+     * - We require debugging. */
+    if (list->device->device_info.device_generated_commands_features_ext.deviceGeneratedCommands)
+    {
+        /* This is implied by specification. */
+        minIndirectCommandsBufferOffsetAlignment = 4;
+        minSequencesCountBufferOffsetAlignment = 4;
+    }
+    else
+    {
+        const VkPhysicalDeviceDeviceGeneratedCommandsPropertiesNV *props =
+                &list->device->device_info.device_generated_commands_properties_nv;
+        minIndirectCommandsBufferOffsetAlignment = props->minIndirectCommandsBufferOffsetAlignment;
+        minSequencesCountBufferOffsetAlignment = props->minSequencesCountBufferOffsetAlignment;
+    }
+
     if ((signature->state_template.dgc.stride != signature->desc.ByteStride && max_command_count > 1) ||
+            (arg_buffer_offset & (minIndirectCommandsBufferOffsetAlignment - 1)) ||
+            (count_buffer && (count_buffer_offset & (minSequencesCountBufferOffsetAlignment - 1))) ||
             patch_args.debug_tag)
     {
         require_patch = true;
@@ -16769,7 +17056,8 @@ static void d3d12_command_list_execute_indirect_state_template_dgc(
         {
             if (FAILED(hr = d3d12_command_allocator_allocate_scratch_memory(list->allocator,
                     VKD3D_SCRATCH_POOL_KIND_DEVICE_STORAGE,
-                    sizeof(uint32_t), sizeof(uint32_t),
+                    sizeof(uint32_t),
+                    minSequencesCountBufferOffsetAlignment,
                     ~0u, &count_allocation)))
             {
                 WARN("Failed to allocate count memory.\n");
@@ -16866,58 +17154,121 @@ static void d3d12_command_list_execute_indirect_state_template_dgc(
         return;
     }
 
-    if (!preprocess_va)
+    if (use_ext_dgc)
     {
-        if (FAILED(hr = d3d12_command_signature_allocate_preprocess_memory_for_list(
-                list, signature, current_pipeline, explicit_preprocess,
-                max_command_count, &preprocess_allocation, &preprocess_size)))
+        if (!preprocess_va)
         {
-            WARN("Failed to allocate preprocess memory.\n");
-            return;
+            if (FAILED(hr = d3d12_command_signature_allocate_preprocess_memory_for_list_ext(
+                    list, signature, current_pipeline, explicit_preprocess,
+                    max_command_count, &preprocess_allocation, &preprocess_size)))
+            {
+                WARN("Failed to allocate preprocess memory.\n");
+                return;
+            }
         }
-    }
 
-    pipeline_info.sType = VK_STRUCTURE_TYPE_GENERATED_COMMANDS_PIPELINE_INFO_EXT;
-    pipeline_info.pNext = NULL;
-    pipeline_info.pipeline = list->current_pipeline;
+        pipeline_info.sType = VK_STRUCTURE_TYPE_GENERATED_COMMANDS_PIPELINE_INFO_EXT;
+        pipeline_info.pNext = NULL;
+        pipeline_info.pipeline = list->current_pipeline;
 
-    memset(&generated_ext, 0, sizeof(generated_ext));
-    generated_ext.sType = VK_STRUCTURE_TYPE_GENERATED_COMMANDS_INFO_EXT;
-    generated_ext.pNext = &pipeline_info;
+        memset(&generated_ext, 0, sizeof(generated_ext));
+        generated_ext.sType = VK_STRUCTURE_TYPE_GENERATED_COMMANDS_INFO_EXT;
+        generated_ext.pNext = &pipeline_info;
 
-    if (signature->pipeline_type == VKD3D_PIPELINE_TYPE_COMPUTE)
-        generated_ext.shaderStages = VK_SHADER_STAGE_COMPUTE_BIT;
-    else
-        generated_ext.shaderStages = list->state->graphics.stage_flags;
+        if (signature->pipeline_type == VKD3D_PIPELINE_TYPE_COMPUTE)
+            generated_ext.shaderStages = VK_SHADER_STAGE_COMPUTE_BIT;
+        else
+            generated_ext.shaderStages = list->state->graphics.stage_flags;
 
-    generated_ext.indirectCommandsLayout = explicit_preprocess || dgc_mode == VKD3D_DGC_MODE_EXECUTE_ONLY ?
-            signature->state_template.dgc.layout_preprocess :
-            signature->state_template.dgc.layout_implicit;
-    generated_ext.preprocessAddress = preprocess_va ? preprocess_va : preprocess_allocation.va;
-    generated_ext.preprocessSize = preprocess_size;
-    generated_ext.maxSequenceCount = max_command_count;
+        generated_ext.indirectCommandsLayout = explicit_preprocess || dgc_mode == VKD3D_DGC_MODE_EXECUTE_ONLY ?
+                signature->state_template.dgc.layout_preprocess_ext :
+                signature->state_template.dgc.layout_implicit_ext;
+        generated_ext.preprocessAddress = preprocess_va ? preprocess_va : preprocess_allocation.va;
+        generated_ext.preprocessSize = preprocess_size;
+        generated_ext.maxSequenceCount = max_command_count;
 
-    if (require_custom_predication)
-    {
-        generated_ext.sequenceCountAddress = predication_allocation.va;
+        if (require_custom_predication)
+        {
+            generated_ext.sequenceCountAddress = predication_allocation.va;
+        }
+        else if (count_buffer)
+        {
+            if (require_patch)
+                generated_ext.sequenceCountAddress = count_allocation.va;
+            else
+                generated_ext.sequenceCountAddress = count_buffer->res.va + count_buffer_offset;
+        }
     }
-    else if (count_buffer)
+    else
     {
-        if (require_patch)
-            generated_ext.sequenceCountAddress = count_allocation.va;
-        else
-            generated_ext.sequenceCountAddress = count_buffer->res.va + count_buffer_offset;
+        if (FAILED(hr = d3d12_command_signature_allocate_preprocess_memory_for_list_nv(
+                list, signature, current_pipeline, explicit_preprocess,
+                max_command_count, &preprocess_allocation, &preprocess_size)))
+        {
+            WARN("Failed to allocate preprocess memory.\n");
+            return;
+        }
+
+        memset(&generated_nv, 0, sizeof(generated_nv));
+        generated_nv.sType = VK_STRUCTURE_TYPE_GENERATED_COMMANDS_INFO_NV;
+        generated_nv.pipeline = list->current_pipeline;
+        generated_nv.pipelineBindPoint = signature->pipeline_type == VKD3D_PIPELINE_TYPE_COMPUTE ?
+                VK_PIPELINE_BIND_POINT_COMPUTE : VK_PIPELINE_BIND_POINT_GRAPHICS;
+        generated_nv.indirectCommandsLayout = explicit_preprocess ?
+                signature->state_template.dgc.layout_preprocess_nv :
+                signature->state_template.dgc.layout_implicit_nv;
+        generated_nv.streamCount = 1;
+        generated_nv.pStreams = &stream;
+        generated_nv.preprocessBuffer = preprocess_allocation.buffer;
+        generated_nv.preprocessOffset = preprocess_allocation.offset;
+        generated_nv.preprocessSize = preprocess_size;
+        generated_nv.sequencesCount = max_command_count;
+
+        if (require_custom_predication)
+        {
+            generated_nv.sequencesCountBuffer = predication_allocation.buffer;
+            generated_nv.sequencesCountOffset = predication_allocation.offset;
+        }
+        else if (count_buffer)
+        {
+            if (require_patch)
+            {
+                generated_nv.sequencesCountBuffer = count_allocation.buffer;
+                generated_nv.sequencesCountOffset = count_allocation.offset;
+            }
+            else
+            {
+                generated_nv.sequencesCountBuffer = count_buffer->res.vk_buffer;
+                generated_nv.sequencesCountOffset = count_buffer->mem.offset + count_buffer_offset;
+            }
+        }
     }
 
     if (require_patch)
     {
-        generated_ext.indirectAddress = stream_allocation.va;
-        generated_ext.indirectAddressSize = max_command_count * signature->state_template.dgc.stride;
+        if (use_ext_dgc)
+        {
+            generated_ext.indirectAddress = stream_allocation.va;
+            generated_ext.indirectAddressSize = max_command_count * signature->state_template.dgc.stride;
+        }
+        else
+        {
+            stream.buffer = stream_allocation.buffer;
+            stream.offset = stream_allocation.offset;
+        }
     }
     else
     {
-        generated_ext.indirectAddress = arg_buffer->res.va + arg_buffer_offset;
-        generated_ext.indirectAddressSize = max_command_count * signature->state_template.dgc.stride;
+        if (use_ext_dgc)
+        {
+            generated_ext.indirectAddress = arg_buffer->res.va + arg_buffer_offset;
+            generated_ext.indirectAddressSize = max_command_count * signature->state_template.dgc.stride;
+        }
+        else
+        {
+            stream.buffer = arg_buffer->res.vk_buffer;
+            stream.offset = arg_buffer->mem.offset + arg_buffer_offset;
+        }
     }
 
     /* If we risk breaking suspend-resume, this will be no-oped out. */
@@ -16936,8 +17287,50 @@ static void d3d12_command_list_execute_indirect_state_template_dgc(
 
         d3d12_command_allocator_allocate_init_post_indirect_command_buffer(list->allocator, list);
 
-        VK_CALL(vkCmdPreprocessGeneratedCommandsEXT(list->cmd.vk_post_indirect_barrier_commands,
-                &generated_ext, list->cmd.vk_command_buffer));
+        if (use_ext_dgc)
+        {
+            VK_CALL(vkCmdPreprocessGeneratedCommandsEXT(list->cmd.vk_post_indirect_barrier_commands,
+                    &generated_ext, list->cmd.vk_command_buffer));
+        }
+        else
+        {
+            VkConditionalRenderingBeginInfoEXT conditional_begin_info;
+            bool explicit_cond_rendering;
+
+            /* With graphics NV_dgc, there are no requirements on bound state, except for pipeline. */
+            /* NV_dgcc however requires that state in recording command buffer matches, but EXT_dgc provides a state cmd. */
+            VK_CALL(vkCmdBindPipeline(list->cmd.vk_post_indirect_barrier_commands,
+                    signature->pipeline_type == VKD3D_PIPELINE_TYPE_COMPUTE ?
+                            VK_PIPELINE_BIND_POINT_COMPUTE : VK_PIPELINE_BIND_POINT_GRAPHICS, current_pipeline));
+
+            if (signature->pipeline_type == VKD3D_PIPELINE_TYPE_COMPUTE)
+            {
+                /* Compute is a bit more stringent, we have to bind all state. */
+                d3d12_command_list_update_descriptors_post_indirect_buffer(list);
+            }
+
+            explicit_cond_rendering = list->predication.enabled_on_command_buffer &&
+                    list->cmd.vk_post_indirect_barrier_commands != list->cmd.vk_command_buffer;
+
+            /* Predication state also has to match. Also useful to nop out explicit preprocess too.
+             * Assumption is that drivers will pull predication state from state command buffer on EXT,
+             * since states have to match. */
+            if (explicit_cond_rendering)
+            {
+                conditional_begin_info.sType = VK_STRUCTURE_TYPE_CONDITIONAL_RENDERING_BEGIN_INFO_EXT;
+                conditional_begin_info.pNext = NULL;
+                conditional_begin_info.buffer = list->predication.vk_buffer;
+                conditional_begin_info.offset = list->predication.vk_buffer_offset;
+                conditional_begin_info.flags = 0;
+                VK_CALL(vkCmdBeginConditionalRenderingEXT(list->cmd.vk_post_indirect_barrier_commands,
+                        &conditional_begin_info));
+            }
+
+            VK_CALL(vkCmdPreprocessGeneratedCommandsNV(list->cmd.vk_post_indirect_barrier_commands, &generated_nv));
+
+            if (explicit_cond_rendering)
+                VK_CALL(vkCmdEndConditionalRenderingEXT(list->cmd.vk_post_indirect_barrier_commands));
+        }
 
         list->cmd.indirect_meta->need_preprocess_barrier = true;
     }
@@ -16962,9 +17355,17 @@ static void d3d12_command_list_execute_indirect_state_template_dgc(
     }
     d3d12_command_list_debug_mark_begin_region(list, tag);
 
-    VK_CALL(vkCmdExecuteGeneratedCommandsEXT(list->cmd.vk_command_buffer,
-            explicit_preprocess || dgc_mode == VKD3D_DGC_MODE_EXECUTE_ONLY ? VK_TRUE : VK_FALSE,
-            &generated_ext));
+    if (use_ext_dgc)
+    {
+        VK_CALL(vkCmdExecuteGeneratedCommandsEXT(list->cmd.vk_command_buffer,
+                explicit_preprocess || dgc_mode == VKD3D_DGC_MODE_EXECUTE_ONLY ? VK_TRUE : VK_FALSE,
+                &generated_ext));
+    }
+    else
+    {
+        VK_CALL(vkCmdExecuteGeneratedCommandsNV(list->cmd.vk_command_buffer,
+                explicit_preprocess ? VK_TRUE : VK_FALSE, &generated_nv));
+    }
 
     d3d12_command_list_debug_mark_end_region(list);
 
@@ -17036,7 +17437,7 @@ void d3d12_command_list_flush_dgc_batch(struct d3d12_command_list *list)
                 if (!d3d12_command_list_update_graphics_pipeline(list, draw->signature->pipeline_type))
                     return;
 
-                if (FAILED(hr = d3d12_command_signature_allocate_preprocess_memory_for_list(
+                if (FAILED(hr = d3d12_command_signature_allocate_preprocess_memory_for_list_ext(
                         list, draw->signature, list->current_pipeline, true,
                         draw->max_command_count, &preprocess_allocation, &draw->preprocess_size)))
                 {
@@ -17143,10 +17544,21 @@ static void STDMETHODCALLTYPE d3d12_command_list_ExecuteIndirect(d3d12_command_l
 
     if (sig_impl->requires_state_template)
     {
-        d3d12_command_list_execute_indirect_state_template_dgc(list, sig_impl,
-                max_command_count,
-                arg_impl, arg_buffer_offset,
-                count_impl, count_buffer_offset, VKD3D_DGC_MODE_APPLICATION_CALL, 0, 0);
+        if (sig_impl->requires_state_template_dgc)
+        {
+            d3d12_command_list_execute_indirect_state_template_dgc(list, sig_impl,
+                    max_command_count,
+                    arg_impl, arg_buffer_offset,
+                    count_impl, count_buffer_offset, VKD3D_DGC_MODE_APPLICATION_CALL, 0, 0);
+        }
+        else if (sig_impl->pipeline_type == VKD3D_PIPELINE_TYPE_COMPUTE)
+        {
+            d3d12_command_list_execute_indirect_state_template_compute(list, sig_impl,
+                    max_command_count,
+                    arg_impl, arg_buffer_offset,
+                    count_impl, count_buffer_offset);
+        }
+
         VKD3D_BREADCRUMB_COMMAND(EXECUTE_INDIRECT_TEMPLATE);
         return;
     }
@@ -20464,8 +20876,11 @@ static VkPipelineStageFlags2 vk_stage_flags_from_d3d12_barrier(struct d3d12_comm
     {
         stages |= VK_PIPELINE_STAGE_2_DRAW_INDIRECT_BIT | VK_PIPELINE_STAGE_2_COMPUTE_SHADER_BIT;
         /* We might use explicit preprocess. */
-        if (list->device->device_info.device_generated_commands_features.deviceGeneratedCommands)
+        if (list->device->device_info.device_generated_commands_features_ext.deviceGeneratedCommands ||
+                list->device->device_info.device_generated_commands_features_nv.deviceGeneratedCommands)
+        {
             stages |= VK_PIPELINE_STAGE_2_COMMAND_PREPROCESS_BIT_EXT;
+        }
     }
 
     if (sync & D3D12_BARRIER_SYNC_COPY_RAYTRACING_ACCELERATION_STRUCTURE)
@@ -20522,7 +20937,8 @@ static VkAccessFlags2 vk_access_flags_from_d3d12_barrier(struct d3d12_command_li
         vk_access |= VK_ACCESS_2_INDIRECT_COMMAND_READ_BIT | VK_ACCESS_2_SHADER_READ_BIT;
 
         /* We might use preprocessing. */
-        if (list->device->device_info.device_generated_commands_features.deviceGeneratedCommands)
+        if (list->device->device_info.device_generated_commands_features_ext.deviceGeneratedCommands ||
+                list->device->device_info.device_generated_commands_features_nv.deviceGeneratedCommands)
             vk_access |= VK_ACCESS_2_COMMAND_PREPROCESS_READ_BIT_EXT;
     }
 
@@ -24805,17 +25221,25 @@ static void d3d12_command_signature_cleanup(struct d3d12_command_signature *sign
 {
     const struct vkd3d_vk_device_procs *vk_procs = &signature->device->vk_procs;
 
-    if (signature->requires_state_template)
+    if (signature->requires_state_template_dgc)
     {
         VK_CALL(vkDestroyBuffer(signature->device->vk_device, signature->state_template.dgc.buffer, NULL));
         vkd3d_free_device_memory(signature->device, &signature->state_template.dgc.memory);
 
-        if (signature->device->device_info.device_generated_commands_features.deviceGeneratedCommands)
+        if (signature->device->device_info.device_generated_commands_features_ext.deviceGeneratedCommands)
         {
             VK_CALL(vkDestroyIndirectCommandsLayoutEXT(signature->device->vk_device,
-                    signature->state_template.dgc.layout_implicit, NULL));
+                    signature->state_template.dgc.layout_implicit_ext, NULL));
             VK_CALL(vkDestroyIndirectCommandsLayoutEXT(signature->device->vk_device,
-                    signature->state_template.dgc.layout_preprocess, NULL));
+                    signature->state_template.dgc.layout_preprocess_ext, NULL));
+        }
+
+        if (signature->device->device_info.device_generated_commands_features_nv.deviceGeneratedCommands)
+        {
+            VK_CALL(vkDestroyIndirectCommandsLayoutNV(signature->device->vk_device,
+                    signature->state_template.dgc.layout_implicit_nv, NULL));
+            VK_CALL(vkDestroyIndirectCommandsLayoutNV(signature->device->vk_device,
+                    signature->state_template.dgc.layout_preprocess_nv, NULL));
         }
     }
 
@@ -24950,25 +25374,66 @@ static HRESULT d3d12_command_signature_init_patch_commands_buffer(struct d3d12_c
     return hr;
 }
 
-static HRESULT d3d12_command_signature_init_indirect_commands_layout(
-        struct d3d12_command_signature *signature,
-        struct d3d12_root_signature *root_signature, struct d3d12_device *device,
-        const VkIndirectCommandsLayoutTokenEXT *tokens, uint32_t token_count,
+static HRESULT d3d12_command_signature_init_indirect_commands_layout_nv(
+        struct d3d12_command_signature *signature, struct d3d12_device *device,
+        const VkIndirectCommandsLayoutTokenNV *tokens, uint32_t token_count,
         uint32_t stream_stride)
 {
     const struct vkd3d_vk_device_procs *vk_procs = &device->vk_procs;
-    VkIndirectCommandsLayoutCreateInfoEXT create_info;
+    VkIndirectCommandsLayoutCreateInfoNV create_info;
     VkResult vr;
 
     memset(&create_info, 0, sizeof(create_info));
-    create_info.sType = VK_STRUCTURE_TYPE_INDIRECT_COMMANDS_LAYOUT_CREATE_INFO_EXT;
+    create_info.sType = VK_STRUCTURE_TYPE_INDIRECT_COMMANDS_LAYOUT_CREATE_INFO_NV;
+    create_info.pipelineBindPoint = signature->pipeline_type == VKD3D_PIPELINE_TYPE_COMPUTE ?
+            VK_PIPELINE_BIND_POINT_COMPUTE : VK_PIPELINE_BIND_POINT_GRAPHICS;
+    create_info.streamCount = 1;
+    create_info.pStreamStrides = &stream_stride;
+    create_info.tokenCount = token_count;
+    create_info.pTokens = tokens;
 
-    switch (signature->pipeline_type)
+    signature->state_template.dgc.stride = stream_stride;
+
+    if (token_count > device->device_info.device_generated_commands_properties_nv.maxIndirectCommandsTokenCount)
     {
-        case VKD3D_PIPELINE_TYPE_COMPUTE:
-            create_info.shaderStages = VK_SHADER_STAGE_COMPUTE_BIT;
-            break;
-        case VKD3D_PIPELINE_TYPE_GRAPHICS:
+        FIXME("Token count %u is too large (max %u).\n",
+                token_count, device->device_info.device_generated_commands_properties_nv.maxIndirectCommandsTokenCount);
+        return E_NOTIMPL;
+    }
+
+    /* Need two separate DGC layouts since if we set EXPLICIT_PREPROCESS, we must use preprocess if the flag is set.
+     * We don't always want to use explicit preprocess (especially when we cannot hoist), so pick the appropriate
+     * layout at ExecuteIndirect time. */
+    vr = VK_CALL(vkCreateIndirectCommandsLayoutNV(device->vk_device, &create_info, NULL,
+            &signature->state_template.dgc.layout_implicit_nv));
+    if (vr != VK_SUCCESS)
+        return hresult_from_vk_result(vr);
+
+    create_info.flags = VK_INDIRECT_COMMANDS_LAYOUT_USAGE_EXPLICIT_PREPROCESS_BIT_NV;
+    vr = VK_CALL(vkCreateIndirectCommandsLayoutNV(device->vk_device, &create_info, NULL,
+            &signature->state_template.dgc.layout_preprocess_nv));
+    return hresult_from_vk_result(vr);
+}
+
+static HRESULT d3d12_command_signature_init_indirect_commands_layout_ext(
+        struct d3d12_command_signature *signature,
+        struct d3d12_root_signature *root_signature, struct d3d12_device *device,
+        const VkIndirectCommandsLayoutTokenEXT *tokens, uint32_t token_count,
+        uint32_t stream_stride)
+{
+    const struct vkd3d_vk_device_procs *vk_procs = &device->vk_procs;
+    VkIndirectCommandsLayoutCreateInfoEXT create_info;
+    VkResult vr;
+
+    memset(&create_info, 0, sizeof(create_info));
+    create_info.sType = VK_STRUCTURE_TYPE_INDIRECT_COMMANDS_LAYOUT_CREATE_INFO_EXT;
+
+    switch (signature->pipeline_type)
+    {
+        case VKD3D_PIPELINE_TYPE_COMPUTE:
+            create_info.shaderStages = VK_SHADER_STAGE_COMPUTE_BIT;
+            break;
+        case VKD3D_PIPELINE_TYPE_GRAPHICS:
             create_info.shaderStages = VK_SHADER_STAGE_ALL_GRAPHICS;
             break;
         case VKD3D_PIPELINE_TYPE_MESH_GRAPHICS:
@@ -24990,10 +25455,10 @@ static HRESULT d3d12_command_signature_init_indirect_commands_layout(
 
     signature->state_template.dgc.stride = stream_stride;
 
-    if (token_count > device->device_info.device_generated_commands_properties.maxIndirectCommandsTokenCount)
+    if (token_count > device->device_info.device_generated_commands_properties_ext.maxIndirectCommandsTokenCount)
     {
         FIXME("Token count %u is too large (max %u).\n",
-                token_count, device->device_info.device_generated_commands_properties.maxIndirectCommandsTokenCount);
+                token_count, device->device_info.device_generated_commands_properties_ext.maxIndirectCommandsTokenCount);
         return E_NOTIMPL;
     }
 
@@ -25001,13 +25466,13 @@ static HRESULT d3d12_command_signature_init_indirect_commands_layout(
      * We don't always want to use explicit preprocess (especially when we cannot hoist), so pick the appropriate
      * layout at ExecuteIndirect time. */
     vr = VK_CALL(vkCreateIndirectCommandsLayoutEXT(device->vk_device, &create_info, NULL,
-            &signature->state_template.dgc.layout_implicit));
+            &signature->state_template.dgc.layout_implicit_ext));
     if (vr != VK_SUCCESS)
         return hresult_from_vk_result(vr);
 
     create_info.flags = VK_INDIRECT_COMMANDS_LAYOUT_USAGE_EXPLICIT_PREPROCESS_BIT_EXT;
     vr = VK_CALL(vkCreateIndirectCommandsLayoutEXT(device->vk_device, &create_info, NULL,
-            &signature->state_template.dgc.layout_preprocess));
+            &signature->state_template.dgc.layout_preprocess_ext));
     return hresult_from_vk_result(vr);
 }
 
@@ -25020,14 +25485,14 @@ static HRESULT d3d12_command_signature_allocate_stream_memory_for_list(
     if (!d3d12_command_allocator_allocate_scratch_memory(list->allocator,
             VKD3D_SCRATCH_POOL_KIND_DEVICE_STORAGE,
             max_command_count * signature->state_template.dgc.stride,
-            sizeof(uint32_t),
+            list->device->device_info.device_generated_commands_properties_nv.minIndirectCommandsBufferOffsetAlignment,
             ~0u, allocation))
         return E_OUTOFMEMORY;
 
     return S_OK;
 }
 
-static HRESULT d3d12_command_signature_allocate_preprocess_memory_for_list(
+static HRESULT d3d12_command_signature_allocate_preprocess_memory_for_list_ext(
         struct d3d12_command_list *list,
         struct d3d12_command_signature *signature, VkPipeline render_pipeline, bool explicit_preprocess,
         uint32_t max_command_count,
@@ -25051,14 +25516,14 @@ static HRESULT d3d12_command_signature_allocate_preprocess_memory_for_list(
     info.pNext = &pipeline_info;
     info.maxSequenceCount = max_command_count;
     info.indirectCommandsLayout = explicit_preprocess ?
-            signature->state_template.dgc.layout_preprocess :
-            signature->state_template.dgc.layout_implicit;
+            signature->state_template.dgc.layout_preprocess_ext :
+            signature->state_template.dgc.layout_implicit_ext;
 
-    if (max_command_count > list->device->device_info.device_generated_commands_properties.maxIndirectSequenceCount)
+    if (max_command_count > list->device->device_info.device_generated_commands_properties_ext.maxIndirectSequenceCount)
     {
         FIXME("max_command_count %u exceeds device limit %u.\n",
                 max_command_count,
-                list->device->device_info.device_generated_commands_properties.maxIndirectSequenceCount);
+                list->device->device_info.device_generated_commands_properties_ext.maxIndirectSequenceCount);
         return E_NOTIMPL;
     }
 
@@ -25078,7 +25543,424 @@ static HRESULT d3d12_command_signature_allocate_preprocess_memory_for_list(
     return S_OK;
 }
 
-static HRESULT d3d12_command_signature_init_state_template_dgc(struct d3d12_command_signature *signature,
+static HRESULT d3d12_command_signature_allocate_preprocess_memory_for_list_nv(
+        struct d3d12_command_list *list,
+        struct d3d12_command_signature *signature, VkPipeline render_pipeline, bool explicit_preprocess,
+        uint32_t max_command_count,
+        struct vkd3d_scratch_allocation *allocation, VkDeviceSize *size)
+{
+    const struct vkd3d_vk_device_procs *vk_procs = &list->device->vk_procs;
+    VkGeneratedCommandsMemoryRequirementsInfoNV info;
+    VkMemoryRequirements2 memory_info;
+    uint32_t alignment;
+
+    memory_info.sType = VK_STRUCTURE_TYPE_MEMORY_REQUIREMENTS_2;
+    memory_info.pNext = NULL;
+
+    info.pipelineBindPoint = signature->pipeline_type == VKD3D_PIPELINE_TYPE_COMPUTE ?
+            VK_PIPELINE_BIND_POINT_COMPUTE : VK_PIPELINE_BIND_POINT_GRAPHICS;
+    info.sType = VK_STRUCTURE_TYPE_GENERATED_COMMANDS_MEMORY_REQUIREMENTS_INFO_NV;
+    info.pNext = NULL;
+    info.maxSequencesCount = max_command_count;
+    info.pipeline = render_pipeline;
+    info.indirectCommandsLayout = explicit_preprocess ?
+            signature->state_template.dgc.layout_preprocess_nv :
+            signature->state_template.dgc.layout_implicit_nv;
+
+    if (max_command_count > list->device->device_info.device_generated_commands_properties_nv.maxIndirectSequenceCount)
+    {
+        FIXME("max_command_count %u exceeds device limit %u.\n",
+                max_command_count,
+                list->device->device_info.device_generated_commands_properties_nv.maxIndirectSequenceCount);
+        return E_NOTIMPL;
+    }
+
+    VK_CALL(vkGetGeneratedCommandsMemoryRequirementsNV(list->device->vk_device, &info, &memory_info));
+
+    alignment = max(memory_info.memoryRequirements.alignment,
+            list->device->device_info.device_generated_commands_properties_nv.minIndirectCommandsBufferOffsetAlignment);
+
+    if (!d3d12_command_allocator_allocate_scratch_memory(list->allocator,
+            VKD3D_SCRATCH_POOL_KIND_INDIRECT_PREPROCESS,
+            memory_info.memoryRequirements.size,
+            alignment,
+            memory_info.memoryRequirements.memoryTypeBits, allocation))
+        return E_OUTOFMEMORY;
+
+    /* Going to assume the memory type is okay ... It's device local after all. */
+    *size = memory_info.memoryRequirements.size;
+    return S_OK;
+}
+
+static HRESULT d3d12_command_signature_init_state_template_compute(struct d3d12_command_signature *signature,
+        const D3D12_COMMAND_SIGNATURE_DESC *desc,
+        struct d3d12_root_signature *root_signature,
+        struct d3d12_device *device)
+{
+    /* Compute templates are simpler, since the only state that can change is
+     * root constants and root descriptors, so we can work around it with some heroics.
+     * The implementation strategy for a non-DGC path is to upload a 256 byte buffer
+     * with default command list root parameter state.
+     * The input is either copied from the buffer directly, or it's read from the indirect buffer and replaces
+     * the default input. This can be done in parallel with 64 threads per dispatch.
+     * Some threads per workgroup will then copy the indirect dispatch parameters
+     * (or clear them to 0 if indirect count needs to mask the dispatch). */
+    const struct vkd3d_shader_root_parameter *root_parameter;
+    const struct vkd3d_shader_root_constant *root_constant;
+    uint32_t root_parameter_index;
+    uint32_t src_offset_words = 0;
+    uint32_t dst_offset_word;
+    unsigned int i, j;
+
+    for (i = 0; i < ARRAY_SIZE(signature->state_template.compute.source_offsets); i++)
+        signature->state_template.compute.source_offsets[i] = -1;
+
+    for (i = 0; i < desc->NumArgumentDescs; i++)
+    {
+        const D3D12_INDIRECT_ARGUMENT_DESC *argument_desc = &desc->pArgumentDescs[i];
+
+        switch (argument_desc->Type)
+        {
+            case D3D12_INDIRECT_ARGUMENT_TYPE_CONSTANT:
+                root_parameter_index = argument_desc->Constant.RootParameterIndex;
+                root_constant = root_signature_get_32bit_constants(root_signature, root_parameter_index);
+
+                dst_offset_word = root_constant->constant_index + argument_desc->Constant.DestOffsetIn32BitValues;
+                for (j = 0; j < argument_desc->Constant.Num32BitValuesToSet; j++, src_offset_words++)
+                    signature->state_template.compute.source_offsets[dst_offset_word + j] = (int32_t)src_offset_words;
+                break;
+
+            case D3D12_INDIRECT_ARGUMENT_TYPE_UNORDERED_ACCESS_VIEW:
+            case D3D12_INDIRECT_ARGUMENT_TYPE_SHADER_RESOURCE_VIEW:
+            case D3D12_INDIRECT_ARGUMENT_TYPE_CONSTANT_BUFFER_VIEW:
+                root_parameter_index = argument_desc->ShaderResourceView.RootParameterIndex;
+                root_parameter = root_signature_get_parameter(root_signature, root_parameter_index);
+
+                if (!(root_signature->root_descriptor_raw_va_mask & (1ull << root_parameter_index)))
+                {
+                    ERR("Root parameter %u is not a raw VA. Cannot implement command signature which updates root descriptor.\n",
+                            root_parameter_index);
+                    return E_NOTIMPL;
+                }
+
+                dst_offset_word = root_parameter->descriptor.raw_va_root_descriptor_index * sizeof(VkDeviceAddress) / sizeof(uint32_t);
+                for (j = 0; j < sizeof(VkDeviceAddress) / sizeof(uint32_t); j++, src_offset_words++)
+                    signature->state_template.compute.source_offsets[dst_offset_word + j] = (int32_t)src_offset_words;
+                break;
+
+            case D3D12_INDIRECT_ARGUMENT_TYPE_DISPATCH:
+                signature->state_template.compute.dispatch_offset_words = src_offset_words;
+                break;
+
+            default:
+                FIXME("Unsupported token type %u.\n", argument_desc->Type);
+                return E_NOTIMPL;
+        }
+    }
+
+    /* No need to build a specialized pipeline here, there is a generic pipeline to handle compute. */
+
+    return S_OK;
+}
+
+static HRESULT d3d12_command_signature_init_state_template_dgc_nv(struct d3d12_command_signature *signature,
+        const D3D12_COMMAND_SIGNATURE_DESC *desc,
+        struct d3d12_root_signature *root_signature,
+        struct d3d12_device *device)
+{
+    const enum vkd3d_patch_command_token *generic_u32_copy_types;
+    const struct vkd3d_shader_root_parameter *root_parameter;
+    const struct d3d12_bind_point_layout *bind_point_layout;
+    const struct vkd3d_shader_root_constant *root_constant;
+    struct vkd3d_patch_command *patch_commands = NULL;
+    VkIndirectCommandsLayoutTokenNV *tokens = NULL;
+    uint32_t required_stride_alignment = 0;
+    VkIndirectCommandsLayoutTokenNV token;
+    uint32_t generic_u32_copy_count;
+    size_t patch_commands_count = 0;
+    uint32_t required_alignment = 0;
+    size_t patch_commands_size = 0;
+    uint32_t root_parameter_index;
+    uint32_t src_word_offset = 0;
+    uint32_t stream_stride = 0;
+    uint32_t dst_word_offset;
+    size_t token_count = 0;
+    size_t token_size = 0;
+    HRESULT hr = S_OK;
+    uint32_t i, j;
+
+    /* Mostly for debug. Lets debug ring report what it is writing easily. */
+    static const enum vkd3d_patch_command_token ibv_types[] =
+    {
+        VKD3D_PATCH_COMMAND_TOKEN_COPY_IBO_VA_LO,
+        VKD3D_PATCH_COMMAND_TOKEN_COPY_IBO_VA_HI,
+        VKD3D_PATCH_COMMAND_TOKEN_COPY_IBO_SIZE,
+        VKD3D_PATCH_COMMAND_TOKEN_COPY_INDEX_FORMAT,
+    };
+
+    static const enum vkd3d_patch_command_token vbv_types[] =
+    {
+        VKD3D_PATCH_COMMAND_TOKEN_COPY_VBO_VA_LO,
+        VKD3D_PATCH_COMMAND_TOKEN_COPY_VBO_VA_HI,
+        VKD3D_PATCH_COMMAND_TOKEN_COPY_VBO_SIZE,
+        VKD3D_PATCH_COMMAND_TOKEN_COPY_VBO_STRIDE,
+    };
+
+    static const enum vkd3d_patch_command_token draw_types[] =
+    {
+        VKD3D_PATCH_COMMAND_TOKEN_COPY_VERTEX_COUNT,
+        VKD3D_PATCH_COMMAND_TOKEN_COPY_INSTANCE_COUNT,
+        VKD3D_PATCH_COMMAND_TOKEN_COPY_FIRST_VERTEX,
+        VKD3D_PATCH_COMMAND_TOKEN_COPY_FIRST_INSTANCE,
+    };
+
+    static const enum vkd3d_patch_command_token draw_indexed_types[] =
+    {
+        VKD3D_PATCH_COMMAND_TOKEN_COPY_INDEX_COUNT,
+        VKD3D_PATCH_COMMAND_TOKEN_COPY_INSTANCE_COUNT,
+        VKD3D_PATCH_COMMAND_TOKEN_COPY_FIRST_INDEX,
+        VKD3D_PATCH_COMMAND_TOKEN_COPY_VERTEX_OFFSET,
+        VKD3D_PATCH_COMMAND_TOKEN_COPY_FIRST_INSTANCE,
+    };
+
+    static const enum vkd3d_patch_command_token draw_mesh_types[] =
+    {
+        VKD3D_PATCH_COMMAND_TOKEN_COPY_MESH_TASKS_X,
+        VKD3D_PATCH_COMMAND_TOKEN_COPY_MESH_TASKS_Y,
+        VKD3D_PATCH_COMMAND_TOKEN_COPY_MESH_TASKS_Z,
+    };
+
+    static const enum vkd3d_patch_command_token va_types[] =
+    {
+        VKD3D_PATCH_COMMAND_TOKEN_COPY_ROOT_VA_LO,
+        VKD3D_PATCH_COMMAND_TOKEN_COPY_ROOT_VA_HI,
+    };
+
+    static const VkIndexType vk_index_types[] = { VK_INDEX_TYPE_UINT32, VK_INDEX_TYPE_UINT16 };
+    static const uint32_t d3d_index_types[] = { DXGI_FORMAT_R32_UINT, DXGI_FORMAT_R16_UINT };
+
+    bind_point_layout = d3d12_root_signature_get_layout(root_signature, signature->pipeline_type);
+
+    for (i = 0; i < desc->NumArgumentDescs; i++)
+    {
+        const D3D12_INDIRECT_ARGUMENT_DESC *argument_desc = &desc->pArgumentDescs[i];
+        memset(&token, 0, sizeof(token));
+        token.sType = VK_STRUCTURE_TYPE_INDIRECT_COMMANDS_LAYOUT_TOKEN_NV;
+        generic_u32_copy_count = 0;
+        dst_word_offset = 0;
+
+        switch (argument_desc->Type)
+        {
+            case D3D12_INDIRECT_ARGUMENT_TYPE_CONSTANT:
+                root_parameter_index = argument_desc->Constant.RootParameterIndex;
+                root_constant = root_signature_get_32bit_constants(root_signature, root_parameter_index);
+
+                if (bind_point_layout->flags & VKD3D_ROOT_SIGNATURE_USE_PUSH_CONSTANT_UNIFORM_BLOCK)
+                {
+                    ERR("Root signature uses push UBO for root parameters, but this feature requires push constant path.\n");
+                    hr = E_NOTIMPL;
+                    goto end;
+                }
+
+                token.tokenType = VK_INDIRECT_COMMANDS_TOKEN_TYPE_PUSH_CONSTANT_NV;
+                token.pushconstantPipelineLayout = bind_point_layout->vk_pipeline_layout;
+                token.pushconstantShaderStageFlags = bind_point_layout->vk_push_stages;
+                token.pushconstantOffset = root_constant->constant_index + argument_desc->Constant.DestOffsetIn32BitValues;
+                token.pushconstantSize = argument_desc->Constant.Num32BitValuesToSet;
+                token.pushconstantOffset *= sizeof(uint32_t);
+                token.pushconstantSize *= sizeof(uint32_t);
+                required_alignment = sizeof(uint32_t);
+
+                stream_stride = align(stream_stride, required_alignment);
+                token.offset = stream_stride;
+                stream_stride += token.pushconstantSize;
+                dst_word_offset = token.offset / sizeof(uint32_t);
+
+                generic_u32_copy_count = argument_desc->Constant.Num32BitValuesToSet;
+                generic_u32_copy_types = NULL;
+                break;
+
+            case D3D12_INDIRECT_ARGUMENT_TYPE_UNORDERED_ACCESS_VIEW:
+            case D3D12_INDIRECT_ARGUMENT_TYPE_SHADER_RESOURCE_VIEW:
+            case D3D12_INDIRECT_ARGUMENT_TYPE_CONSTANT_BUFFER_VIEW:
+                root_parameter_index = argument_desc->ShaderResourceView.RootParameterIndex;
+                root_parameter = root_signature_get_parameter(root_signature, root_parameter_index);
+
+                if (bind_point_layout->flags & VKD3D_ROOT_SIGNATURE_USE_PUSH_CONSTANT_UNIFORM_BLOCK)
+                {
+                    ERR("Root signature uses push UBO for root parameters, but this feature requires push constant path.\n");
+                    hr = E_NOTIMPL;
+                    goto end;
+                }
+
+                if (!(root_signature->root_descriptor_raw_va_mask & (1ull << root_parameter_index)))
+                {
+                    ERR("Root parameter %u is not a raw VA. Cannot implement command signature which updates root descriptor.\n",
+                            root_parameter_index);
+                    hr = E_NOTIMPL;
+                    goto end;
+                }
+
+                token.tokenType = VK_INDIRECT_COMMANDS_TOKEN_TYPE_PUSH_CONSTANT_NV;
+                token.pushconstantPipelineLayout = bind_point_layout->vk_pipeline_layout;
+                token.pushconstantShaderStageFlags = bind_point_layout->vk_push_stages;
+                token.pushconstantOffset = root_parameter->descriptor.raw_va_root_descriptor_index * sizeof(VkDeviceAddress);
+                token.pushconstantSize = sizeof(VkDeviceAddress);
+                required_alignment = sizeof(uint32_t);
+
+                stream_stride = align(stream_stride, required_alignment);
+                token.offset = stream_stride;
+                stream_stride += token.pushconstantSize;
+                dst_word_offset = token.offset / sizeof(uint32_t);
+
+                /* Simply patch by copying U32s. Need to handle unaligned U32s since everything is tightly packed. */
+                generic_u32_copy_count = sizeof(VkDeviceAddress) / sizeof(uint32_t);
+                generic_u32_copy_types = va_types;
+                break;
+
+            case D3D12_INDIRECT_ARGUMENT_TYPE_VERTEX_BUFFER_VIEW:
+                token.tokenType = VK_INDIRECT_COMMANDS_TOKEN_TYPE_VERTEX_BUFFER_NV;
+                token.vertexBindingUnit = argument_desc->VertexBuffer.Slot;
+                token.vertexDynamicStride = VK_TRUE;
+
+                /* If device exposes 4 byte alignment of the indirect command buffer, we can
+                 * pack VA at sub-scalar alignment. */
+                required_alignment = min(
+                        device->device_info.device_generated_commands_properties_nv.minIndirectCommandsBufferOffsetAlignment,
+                        sizeof(VkDeviceAddress));
+
+                stream_stride = align(stream_stride, required_alignment);
+                token.offset = stream_stride;
+                stream_stride += sizeof(VkBindVertexBufferIndirectCommandNV);
+                dst_word_offset = token.offset / sizeof(uint32_t);
+
+                /* The VBV indirect layout is the same as DX, so just copy the U32s. */
+                generic_u32_copy_count = sizeof(D3D12_VERTEX_BUFFER_VIEW) / sizeof(uint32_t);
+                generic_u32_copy_types = vbv_types;
+                break;
+
+            case D3D12_INDIRECT_ARGUMENT_TYPE_INDEX_BUFFER_VIEW:
+                token.tokenType = VK_INDIRECT_COMMANDS_TOKEN_TYPE_INDEX_BUFFER_NV;
+                token.indexTypeCount = ARRAY_SIZE(vk_index_types);
+                token.pIndexTypeValues = d3d_index_types;
+                token.pIndexTypes = vk_index_types;
+
+                /* If device exposes 4 byte alignment of the indirect command buffer, we can
+                 * pack VA at sub-scalar alignment. */
+                required_alignment = min(
+                        device->device_info.device_generated_commands_properties_nv.minIndirectCommandsBufferOffsetAlignment,
+                        sizeof(VkDeviceAddress));
+
+                stream_stride = align(stream_stride, required_alignment);
+                token.offset = stream_stride;
+                stream_stride += sizeof(VkBindVertexBufferIndirectCommandNV);
+                dst_word_offset = token.offset / sizeof(uint32_t);
+
+                vkd3d_array_reserve((void**)&patch_commands, &patch_commands_size,
+                        patch_commands_count + sizeof(D3D12_INDEX_BUFFER_VIEW) / sizeof(uint32_t),
+                        sizeof(*patch_commands));
+
+                for (j = 0; j < 4; j++)
+                {
+                    patch_commands[patch_commands_count].token = ibv_types[j];
+                    patch_commands[patch_commands_count].src_offset = src_word_offset++;
+                    patch_commands[patch_commands_count].dst_offset = dst_word_offset++;
+                    patch_commands_count++;
+                }
+                break;
+
+            case D3D12_INDIRECT_ARGUMENT_TYPE_DRAW:
+                token.tokenType = VK_INDIRECT_COMMANDS_TOKEN_TYPE_DRAW_NV;
+                required_alignment = sizeof(uint32_t);
+                stream_stride = align(stream_stride, required_alignment);
+                token.offset = stream_stride;
+                stream_stride += sizeof(VkDrawIndirectCommand);
+                dst_word_offset = token.offset / sizeof(uint32_t);
+                generic_u32_copy_count = sizeof(VkDrawIndirectCommand) / sizeof(uint32_t);
+                generic_u32_copy_types = draw_types;
+                break;
+
+            case D3D12_INDIRECT_ARGUMENT_TYPE_DRAW_INDEXED:
+                token.tokenType = VK_INDIRECT_COMMANDS_TOKEN_TYPE_DRAW_INDEXED_NV;
+                required_alignment = sizeof(uint32_t);
+                stream_stride = align(stream_stride, required_alignment);
+                token.offset = stream_stride;
+                stream_stride += sizeof(VkDrawIndexedIndirectCommand);
+                dst_word_offset = token.offset / sizeof(uint32_t);
+                generic_u32_copy_count = sizeof(VkDrawIndexedIndirectCommand) / sizeof(uint32_t);
+                generic_u32_copy_types = draw_indexed_types;
+                break;
+
+            case D3D12_INDIRECT_ARGUMENT_TYPE_DISPATCH_MESH:
+                token.tokenType = VK_INDIRECT_COMMANDS_TOKEN_TYPE_DRAW_MESH_TASKS_NV;
+                required_alignment = sizeof(uint32_t);
+                stream_stride = align(stream_stride, required_alignment);
+                token.offset = stream_stride;
+                stream_stride += sizeof(VkDrawMeshTasksIndirectCommandEXT);
+                dst_word_offset = token.offset / sizeof(uint32_t);
+                generic_u32_copy_count = sizeof(VkDrawMeshTasksIndirectCommandEXT) / sizeof(uint32_t);
+                generic_u32_copy_types = draw_mesh_types;
+                break;
+
+            case D3D12_INDIRECT_ARGUMENT_TYPE_DISPATCH:
+                token.tokenType = VK_INDIRECT_COMMANDS_TOKEN_TYPE_DISPATCH_NV;
+                required_alignment = sizeof(uint32_t);
+                stream_stride = align(stream_stride, required_alignment);
+                token.offset = stream_stride;
+                stream_stride += sizeof(VkDispatchIndirectCommand);
+                dst_word_offset = token.offset / sizeof(uint32_t);
+                /* TODO: Rebase on top of debug-ring-indirect. */
+                generic_u32_copy_count = 0;
+                generic_u32_copy_types = NULL;
+                break;
+
+            default:
+                FIXME("Unsupported token type %u.\n", argument_desc->Type);
+                hr = E_NOTIMPL;
+                goto end;
+        }
+
+        vkd3d_array_reserve((void**)&tokens, &token_size, token_count + 1, sizeof(*tokens));
+        tokens[token_count++] = token;
+
+        if (generic_u32_copy_count)
+        {
+            vkd3d_array_reserve((void**)&patch_commands, &patch_commands_size,
+                    patch_commands_count + generic_u32_copy_count,
+                    sizeof(*patch_commands));
+
+            /* Simply patch by copying U32s. */
+            for (j = 0; j < generic_u32_copy_count; j++, patch_commands_count++)
+            {
+                patch_commands[patch_commands_count].token =
+                        generic_u32_copy_types ? generic_u32_copy_types[j] : VKD3D_PATCH_COMMAND_TOKEN_COPY_CONST_U32;
+                patch_commands[patch_commands_count].src_offset = src_word_offset++;
+                patch_commands[patch_commands_count].dst_offset = dst_word_offset++;
+            }
+        }
+
+        /* Required alignment is scalar alignment rules, i.e. maximum individual alignment requirement. */
+        required_stride_alignment = max(required_stride_alignment, required_alignment);
+    }
+
+    stream_stride = max(stream_stride, desc->ByteStride);
+    stream_stride = align(stream_stride, required_stride_alignment);
+
+    if (FAILED(hr = d3d12_command_signature_init_patch_commands_buffer(signature, device, patch_commands, patch_commands_count)))
+        goto end;
+    if (FAILED(hr = d3d12_command_signature_init_indirect_commands_layout_nv(signature, device, tokens, token_count, stream_stride)))
+        goto end;
+    if (FAILED(hr = vkd3d_meta_get_execute_indirect_pipeline(&device->meta_ops, patch_commands_count,
+            &signature->state_template.dgc.pipeline)))
+        goto end;
+
+end:
+    vkd3d_free(tokens);
+    vkd3d_free(patch_commands);
+    return hr;
+}
+
+static HRESULT d3d12_command_signature_init_state_template_dgc_ext(struct d3d12_command_signature *signature,
         const D3D12_COMMAND_SIGNATURE_DESC *desc,
         struct d3d12_root_signature *root_signature,
         struct d3d12_device *device)
@@ -25383,7 +26265,7 @@ static HRESULT d3d12_command_signature_init_state_template_dgc(struct d3d12_comm
         if (FAILED(hr = d3d12_command_signature_init_patch_commands_buffer(signature, device, patch_commands, patch_commands_count)))
             goto end;
 
-    if (FAILED(hr = d3d12_command_signature_init_indirect_commands_layout(signature, root_signature, device, tokens, token_count, stream_stride)))
+    if (FAILED(hr = d3d12_command_signature_init_indirect_commands_layout_ext(signature, root_signature, device, tokens, token_count, stream_stride)))
         goto end;
 
     if (patch_commands_count)
@@ -25559,14 +26441,26 @@ HRESULT d3d12_command_signature_create(struct d3d12_device *device, struct d3d12
 
     if ((object->requires_state_template = requires_state_template))
     {
-        if (!device->device_info.device_generated_commands_features.deviceGeneratedCommands)
+        if ((pipeline_type == VKD3D_PIPELINE_TYPE_GRAPHICS || pipeline_type == VKD3D_PIPELINE_TYPE_MESH_GRAPHICS) &&
+                !device->device_info.device_generated_commands_features_nv.deviceGeneratedCommands &&
+                !device->device_info.device_generated_commands_features_ext.deviceGeneratedCommands)
         {
             FIXME("Device generated commands is not supported by implementation.\n");
             object->requires_state_template = false;
             goto out;
         }
-
-        if (pipeline_type == VKD3D_PIPELINE_TYPE_RAY_TRACING)
+        else if (pipeline_type == VKD3D_PIPELINE_TYPE_COMPUTE)
+        {
+            if (!device->device_info.device_generated_commands_compute_features_nv.deviceGeneratedCompute &&
+                    !device->device_info.device_generated_commands_features_ext.deviceGeneratedCommands &&
+                    !(device->bindless_state.flags & VKD3D_FORCE_COMPUTE_ROOT_PARAMETERS_PUSH_UBO))
+            {
+                FIXME("State template is required for compute, but VKD3D_CONFIG_FLAG_REQUIRES_COMPUTE_INDIRECT_TEMPLATES is not enabled.\n");
+                object->requires_state_template = false;
+                goto out;
+            }
+        }
+        else if (pipeline_type == VKD3D_PIPELINE_TYPE_RAY_TRACING)
         {
             /* Very similar idea as indirect compute would be. */
             FIXME("State template is required for indirect ray tracing, but it is unimplemented.\n");
@@ -25574,9 +26468,25 @@ HRESULT d3d12_command_signature_create(struct d3d12_device *device, struct d3d12
             goto out;
         }
 
-        if (device->device_info.device_generated_commands_features.deviceGeneratedCommands)
-            if (FAILED(hr = d3d12_command_signature_init_state_template_dgc(object, desc, root_signature, device)))
+        if (device->device_info.device_generated_commands_features_ext.deviceGeneratedCommands)
+        {
+            if (FAILED(hr = d3d12_command_signature_init_state_template_dgc_ext(object, desc, root_signature, device)))
                 goto err;
+            object->requires_state_template_dgc = true;
+        }
+        else if (pipeline_type == VKD3D_PIPELINE_TYPE_GRAPHICS || pipeline_type == VKD3D_PIPELINE_TYPE_MESH_GRAPHICS ||
+                (pipeline_type == VKD3D_PIPELINE_TYPE_COMPUTE &&
+                        device->device_info.device_generated_commands_compute_features_nv.deviceGeneratedCompute))
+        {
+            if (FAILED(hr = d3d12_command_signature_init_state_template_dgc_nv(object, desc, root_signature, device)))
+                goto err;
+            object->requires_state_template_dgc = true;
+        }
+        else if (pipeline_type == VKD3D_PIPELINE_TYPE_COMPUTE)
+        {
+            if (FAILED(hr = d3d12_command_signature_init_state_template_compute(object, desc, root_signature, device)))
+                goto err;
+        }
 
         /* Heuristic. If game uses fancy execute indirect we're more inclined to split command buffers
          * for optimal reordering. */
diff --git a/libs/vkd3d/device.c b/libs/vkd3d/device.c
index 63a82c3..3c029ec 100644
--- a/libs/vkd3d/device.c
+++ b/libs/vkd3d/device.c
@@ -147,8 +147,10 @@ static const struct vkd3d_optional_extension_info optional_device_extensions[] =
     VK_EXTENSION_DISABLE_COND(NVX_IMAGE_VIEW_HANDLE, NVX_image_view_handle, VKD3D_CONFIG_FLAG_NO_NVX),
     VK_EXTENSION(NV_COMPUTE_SHADER_DERIVATIVES, NV_compute_shader_derivatives),
     VK_EXTENSION_COND(NV_DEVICE_DIAGNOSTIC_CHECKPOINTS, NV_device_diagnostic_checkpoints, VKD3D_CONFIG_FLAG_BREADCRUMBS | VKD3D_CONFIG_FLAG_BREADCRUMBS_TRACE),
+    VK_EXTENSION(NV_DEVICE_GENERATED_COMMANDS, NV_device_generated_commands),
     VK_EXTENSION(NV_SHADER_SUBGROUP_PARTITIONED, NV_shader_subgroup_partitioned),
     VK_EXTENSION(NV_MEMORY_DECOMPRESSION, NV_memory_decompression),
+    VK_EXTENSION(NV_DEVICE_GENERATED_COMMANDS_COMPUTE, NV_device_generated_commands_compute),
     VK_EXTENSION_VERSION(NV_LOW_LATENCY_2, NV_low_latency2, 2),
     VK_EXTENSION(NV_RAW_ACCESS_CHAINS, NV_raw_access_chains),
     VK_EXTENSION(NV_COOPERATIVE_MATRIX_2, NV_cooperative_matrix2),
@@ -610,7 +612,8 @@ static const struct vkd3d_instance_application_meta application_override[] = {
     { VKD3D_STRING_COMPARE_EXACT, "HaloInfinite.exe",
             VKD3D_CONFIG_FLAG_FORCE_RAW_VA_CBV |
             VKD3D_CONFIG_FLAG_USE_HOST_IMPORT_FALLBACK | VKD3D_CONFIG_FLAG_PREALLOCATE_SRV_MIP_CLAMPS |
-            VKD3D_CONFIG_FLAG_NO_UPLOAD_HVV, 0 },
+            VKD3D_CONFIG_FLAG_REQUIRES_COMPUTE_INDIRECT_TEMPLATES | VKD3D_CONFIG_FLAG_NO_UPLOAD_HVV |
+            VKD3D_CONFIG_FLAG_DISABLE_DGCC, 0 },
     /* (1182900) Workaround amdgpu kernel bug with host memory import and concurrent submissions. */
     { VKD3D_STRING_COMPARE_EXACT, "APlagueTaleRequiem_x64.exe",
             VKD3D_CONFIG_FLAG_USE_HOST_IMPORT_FALLBACK | VKD3D_CONFIG_FLAG_DISABLE_UAV_COMPRESSION, 0 },
@@ -655,7 +658,7 @@ static const struct vkd3d_instance_application_meta application_override[] = {
     { VKD3D_STRING_COMPARE_EXACT, "RDR.exe", VKD3D_CONFIG_FLAG_NO_UPLOAD_HVV, 0 },
     /* Starfield (1716740) */
     { VKD3D_STRING_COMPARE_EXACT, "Starfield.exe",
-            VKD3D_CONFIG_FLAG_HUGE_NV_DGC_BUFFERS | VKD3D_CONFIG_FLAG_REJECT_PADDED_SMALL_RESOURCE_ALIGNMENT, 0 },
+            VKD3D_CONFIG_FLAG_REQUIRES_COMPUTE_INDIRECT_TEMPLATES | VKD3D_CONFIG_FLAG_REJECT_PADDED_SMALL_RESOURCE_ALIGNMENT, 0 },
     /* Persona 3 Reload (2161700). Enables RT by default on Deck and does not run acceptably for a verified title. */
     { VKD3D_STRING_COMPARE_EXACT, "P3R.exe", 0, 0, VKD3D_APPLICATION_FEATURE_NO_DEFAULT_DXR_ON_DECK },
     /* Basically never bothers doing initial transitions.
@@ -1282,6 +1285,7 @@ static const struct vkd3d_debug_option vkd3d_config_options[] =
     {"preallocate_srv_mip_clamps", VKD3D_CONFIG_FLAG_PREALLOCATE_SRV_MIP_CLAMPS},
     {"force_initial_transition", VKD3D_CONFIG_FLAG_FORCE_INITIAL_TRANSITION},
     {"breadcrumbs_trace", VKD3D_CONFIG_FLAG_BREADCRUMBS | VKD3D_CONFIG_FLAG_BREADCRUMBS_TRACE},
+    {"requires_compute_indirect_templates", VKD3D_CONFIG_FLAG_REQUIRES_COMPUTE_INDIRECT_TEMPLATES},
     {"skip_driver_workarounds", VKD3D_CONFIG_FLAG_SKIP_DRIVER_WORKAROUNDS},
     {"enable_experimental_features", VKD3D_CONFIG_FLAG_ENABLE_EXPERIMENTAL_FEATURES},
     {"reject_padded_small_resource_alignment", VKD3D_CONFIG_FLAG_REJECT_PADDED_SMALL_RESOURCE_ALIGNMENT},
@@ -1837,6 +1841,17 @@ static void vkd3d_physical_device_info_apply_workarounds(struct vkd3d_physical_d
 
     if (!(vkd3d_config_flags & VKD3D_CONFIG_FLAG_SKIP_DRIVER_WORKAROUNDS))
     {
+        if (info->vulkan_1_2_properties.driverID == VK_DRIVER_ID_NVIDIA_PROPRIETARY &&
+                device->vk_info.NV_device_generated_commands_compute &&
+                (vkd3d_config_flags & VKD3D_CONFIG_FLAG_DISABLE_DGCC))
+        {
+            device->vk_info.NV_device_generated_commands_compute = false;
+            device->vk_info.EXT_device_generated_commands = false;
+            device->device_info.device_generated_commands_compute_features_nv.deviceGeneratedCompute = VK_FALSE;
+            device->device_info.device_generated_commands_features_ext.deviceGeneratedCommands = VK_FALSE;
+            WARN("Disabling DGCC due to config flag.\n");
+        }
+
         /* Two known bugs in the wild:
          * - presentID = 0 handling when toggling present mode is broken.
          * - swapchain fence is not enough to avoid DEVICE_LOST when resizing swapchain.
@@ -2051,12 +2066,12 @@ static void vkd3d_physical_device_info_init(struct vkd3d_physical_device_info *i
 
     if (vulkan_info->EXT_device_generated_commands)
     {
-        info->device_generated_commands_features.sType =
+        info->device_generated_commands_features_ext.sType =
                 VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_DEVICE_GENERATED_COMMANDS_FEATURES_EXT;
-        info->device_generated_commands_properties.sType =
+        info->device_generated_commands_properties_ext.sType =
                 VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_DEVICE_GENERATED_COMMANDS_PROPERTIES_EXT;
-        vk_prepend_struct(&info->features2, &info->device_generated_commands_features);
-        vk_prepend_struct(&info->properties2, &info->device_generated_commands_properties);
+        vk_prepend_struct(&info->features2, &info->device_generated_commands_features_ext);
+        vk_prepend_struct(&info->properties2, &info->device_generated_commands_properties_ext);
     }
 
     if (vulkan_info->EXT_robustness2)
@@ -2171,6 +2186,23 @@ static void vkd3d_physical_device_info_init(struct vkd3d_physical_device_info *i
         vk_prepend_struct(&info->features2, &info->barycentric_features_khr);
     }
 
+    if (vulkan_info->NV_device_generated_commands)
+    {
+        info->device_generated_commands_features_nv.sType =
+                VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_DEVICE_GENERATED_COMMANDS_FEATURES_NV;
+        info->device_generated_commands_properties_nv.sType =
+                VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_DEVICE_GENERATED_COMMANDS_PROPERTIES_NV;
+        vk_prepend_struct(&info->features2, &info->device_generated_commands_features_nv);
+        vk_prepend_struct(&info->properties2, &info->device_generated_commands_properties_nv);
+    }
+
+    if (vulkan_info->NV_device_generated_commands_compute)
+    {
+        info->device_generated_commands_compute_features_nv.sType =
+                VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_DEVICE_GENERATED_COMMANDS_COMPUTE_FEATURES_NV;
+        vk_prepend_struct(&info->features2, &info->device_generated_commands_compute_features_nv);
+    }
+
     if (vulkan_info->EXT_shader_image_atomic_int64)
     {
         info->shader_image_atomic_int64_features.sType =
@@ -3266,11 +3298,13 @@ static HRESULT vkd3d_init_device_caps(struct d3d12_device *device,
     physical_device_info->extended_dynamic_state2_features.extendedDynamicState2LogicOp = VK_FALSE;
 
     /* Unneeded. */
-    physical_device_info->device_generated_commands_features.dynamicGeneratedPipelineLayout = VK_FALSE;
+    physical_device_info->device_generated_commands_compute_features_nv.deviceGeneratedComputeCaptureReplay = VK_FALSE;
+    physical_device_info->device_generated_commands_compute_features_nv.deviceGeneratedComputePipelines = VK_FALSE;
+    physical_device_info->device_generated_commands_features_ext.dynamicGeneratedPipelineLayout = VK_FALSE;
 
     {
         const VkShaderStageFlags supported_stages =
-                physical_device_info->device_generated_commands_properties.supportedIndirectCommandsShaderStages;
+                physical_device_info->device_generated_commands_properties_ext.supportedIndirectCommandsShaderStages;
         VkShaderStageFlags required_stages = VK_SHADER_STAGE_ALL_GRAPHICS | VK_SHADER_STAGE_COMPUTE_BIT;
         if (physical_device_info->mesh_shader_features.meshShader)
             required_stages |= VK_SHADER_STAGE_MESH_BIT_EXT | VK_SHADER_STAGE_TASK_BIT_EXT;
@@ -3279,11 +3313,17 @@ static HRESULT vkd3d_init_device_caps(struct d3d12_device *device,
         {
             /* Can be relaxed if needed in the wild, but in principle, everything needs to be supported
              * to be D3D12 compliant. */
-            INFO("Not all relevant pipeline stages are supported by EXT_dgc. Skipping.\n");
-            physical_device_info->device_generated_commands_features.deviceGeneratedCommands = VK_FALSE;
+            INFO("Not all relevant pipeline stages are supported by EXT_dgc. Skipping EXT.\n");
+            physical_device_info->device_generated_commands_features_ext.deviceGeneratedCommands = VK_FALSE;
         }
     }
 
+    if (physical_device_info->device_generated_commands_features_ext.deviceGeneratedCommands)
+    {
+        physical_device_info->device_generated_commands_features_nv.deviceGeneratedCommands = VK_FALSE;
+        physical_device_info->device_generated_commands_compute_features_nv.deviceGeneratedCompute = VK_FALSE;
+    }
+
     if (!physical_device_info->vulkan_1_2_properties.robustBufferAccessUpdateAfterBind)
     {
         /* Generally, we cannot enable robustness if this is not supported,
@@ -4176,7 +4216,7 @@ static HRESULT d3d12_device_create_scratch_buffer(struct d3d12_device *device, e
         alloc_info.heap_flags = D3D12_HEAP_FLAG_ALLOW_ONLY_BUFFERS | D3D12_HEAP_FLAG_CREATE_NOT_ZEROED;
         alloc_info.flags = VKD3D_ALLOCATION_FLAG_GLOBAL_BUFFER | VKD3D_ALLOCATION_FLAG_INTERNAL_SCRATCH;
         alloc_info.vk_memory_priority = vkd3d_convert_to_vk_prio(D3D12_RESIDENCY_PRIORITY_NORMAL);
-        if (device->device_info.device_generated_commands_features.deviceGeneratedCommands)
+        if (device->device_info.device_generated_commands_features_ext.deviceGeneratedCommands)
         {
             /* this flag cannot be used with the existing buffer heaps */
             alloc_info.explicit_global_buffer_usage =
@@ -9697,7 +9737,7 @@ static void d3d12_device_caps_init_feature_options21(struct d3d12_device *device
     /* Enable WGs in test suite to avoid bitrot. */
     options21->WorkGraphsTier = d3d12_device_supports_workgraphs(device) ?
             D3D12_WORK_GRAPHS_TIER_1_0 : D3D12_WORK_GRAPHS_TIER_NOT_SUPPORTED;
-    options21->ExecuteIndirectTier = device->device_info.device_generated_commands_features.deviceGeneratedCommands ?
+    options21->ExecuteIndirectTier = device->device_info.device_generated_commands_features_ext.deviceGeneratedCommands ?
             D3D12_EXECUTE_INDIRECT_TIER_1_1 : D3D12_EXECUTE_INDIRECT_TIER_1_0;
     options21->SampleCmpGradientAndBiasSupported = device->d3d12_caps.max_shader_model >= D3D_SHADER_MODEL_6_8 &&
             device->d3d12_caps.options14.AdvancedTextureOpsSupported;
@@ -10473,8 +10513,9 @@ static void vkd3d_scratch_pool_init(struct d3d12_device *device)
 
     if (device->device_info.vulkan_1_2_properties.driverID == VK_DRIVER_ID_NVIDIA_PROPRIETARY)
     {
-        if ((vkd3d_config_flags & VKD3D_CONFIG_FLAG_HUGE_NV_DGC_BUFFERS) &&
-            device->device_info.device_generated_commands_features.deviceGeneratedCommands)
+        if ((vkd3d_config_flags & VKD3D_CONFIG_FLAG_REQUIRES_COMPUTE_INDIRECT_TEMPLATES) &&
+            (device->device_info.device_generated_commands_compute_features_nv.deviceGeneratedCompute ||
+             device->device_info.device_generated_commands_features_ext.deviceGeneratedCommands))
         {
             /* DGCC preprocess buffers are gigantic on NV. Starfield requires 27 MB for 4096 dispatches ... */
             device->scratch_pools[VKD3D_SCRATCH_POOL_KIND_INDIRECT_PREPROCESS].block_size =
diff --git a/libs/vkd3d/meson.build b/libs/vkd3d/meson.build
index 477b366..5773f78 100644
--- a/libs/vkd3d/meson.build
+++ b/libs/vkd3d/meson.build
@@ -35,6 +35,7 @@ vkd3d_shaders =[
   'shaders/cs_execute_indirect_patch.comp',
   'shaders/cs_execute_indirect_patch_debug_ring.comp',
   'shaders/cs_execute_indirect_multi_dispatch.comp',
+  'shaders/cs_execute_indirect_multi_dispatch_state.comp',
 
   'shaders/cs_sampler_feedback_decode_buffer_min_mip.comp',
   'shaders/fs_sampler_feedback_decode_image_min_mip.frag',
diff --git a/libs/vkd3d/meta.c b/libs/vkd3d/meta.c
index 770efda..ce5b3f6 100644
--- a/libs/vkd3d/meta.c
+++ b/libs/vkd3d/meta.c
@@ -1530,8 +1530,12 @@ static void vkd3d_multi_dispatch_indirect_ops_cleanup(
     const struct vkd3d_vk_device_procs *vk_procs = &device->vk_procs;
     VK_CALL(vkDestroyPipeline(device->vk_device,
             meta_multi_dispatch_indirect_ops->vk_multi_dispatch_indirect_pipeline, NULL));
+    VK_CALL(vkDestroyPipeline(device->vk_device,
+            meta_multi_dispatch_indirect_ops->vk_multi_dispatch_indirect_state_pipeline, NULL));
     VK_CALL(vkDestroyPipelineLayout(device->vk_device,
             meta_multi_dispatch_indirect_ops->vk_multi_dispatch_indirect_layout, NULL));
+    VK_CALL(vkDestroyPipelineLayout(device->vk_device,
+            meta_multi_dispatch_indirect_ops->vk_multi_dispatch_indirect_state_layout, NULL));
 }
 
 static HRESULT vkd3d_multi_dispatch_indirect_ops_init(
@@ -1550,12 +1554,24 @@ static HRESULT vkd3d_multi_dispatch_indirect_ops_init(
             &push_constant_range, &meta_multi_dispatch_indirect_ops->vk_multi_dispatch_indirect_layout)) < 0)
         goto fail;
 
+    push_constant_range.size = sizeof(struct vkd3d_multi_dispatch_indirect_state_args);
+
+    if ((vr = vkd3d_meta_create_pipeline_layout(device, 0, NULL, 1,
+            &push_constant_range, &meta_multi_dispatch_indirect_ops->vk_multi_dispatch_indirect_state_layout)) < 0)
+        goto fail;
+
     if ((vr = vkd3d_meta_create_compute_pipeline(device,
             sizeof(cs_execute_indirect_multi_dispatch), cs_execute_indirect_multi_dispatch,
             meta_multi_dispatch_indirect_ops->vk_multi_dispatch_indirect_layout, NULL, true, NULL,
             &meta_multi_dispatch_indirect_ops->vk_multi_dispatch_indirect_pipeline)) < 0)
         goto fail;
 
+    if ((vr = vkd3d_meta_create_compute_pipeline(device,
+            sizeof(cs_execute_indirect_multi_dispatch_state), cs_execute_indirect_multi_dispatch_state,
+            meta_multi_dispatch_indirect_ops->vk_multi_dispatch_indirect_state_layout, NULL, true, NULL,
+            &meta_multi_dispatch_indirect_ops->vk_multi_dispatch_indirect_state_pipeline)) < 0)
+        goto fail;
+
     return S_OK;
 
 fail:
@@ -1665,6 +1681,13 @@ void vkd3d_meta_get_multi_dispatch_indirect_pipeline(struct vkd3d_meta_ops *meta
     info->vk_pipeline_layout = meta_ops->multi_dispatch_indirect.vk_multi_dispatch_indirect_layout;
 }
 
+void vkd3d_meta_get_multi_dispatch_indirect_state_pipeline(struct vkd3d_meta_ops *meta_ops,
+        struct vkd3d_multi_dispatch_indirect_info *info)
+{
+    info->vk_pipeline = meta_ops->multi_dispatch_indirect.vk_multi_dispatch_indirect_state_pipeline;
+    info->vk_pipeline_layout = meta_ops->multi_dispatch_indirect.vk_multi_dispatch_indirect_state_layout;
+}
+
 static HRESULT vkd3d_execute_indirect_ops_init(struct vkd3d_execute_indirect_ops *meta_indirect_ops,
         struct d3d12_device *device)
 {
diff --git a/libs/vkd3d/shaders/cs_execute_indirect_multi_dispatch_state.comp b/libs/vkd3d/shaders/cs_execute_indirect_multi_dispatch_state.comp
new file mode 100644
index 0000000..acb8f5f
--- /dev/null
+++ b/libs/vkd3d/shaders/cs_execute_indirect_multi_dispatch_state.comp
@@ -0,0 +1,85 @@
+#version 450
+#extension GL_EXT_buffer_reference : require
+#extension GL_EXT_buffer_reference_uvec2 : require
+
+// Always emit a full 256 byte payload for each UBO. We have to assume worst case alignment anyways,
+// little point in trying to be smart about it.
+layout(local_size_x = 64) in;
+
+layout(buffer_reference_align = 4, std430, buffer_reference) readonly buffer Indirect
+{
+	uint values[];
+};
+
+layout(buffer_reference_align = 4, std430, buffer_reference) writeonly buffer IndirectOutput
+{
+	uint values[];
+};
+
+layout(buffer_reference_align = 4, std430, buffer_reference) readonly buffer IndirectCount
+{
+	uint value;
+};
+
+struct RootParameterBlock
+{
+	uint values[64];
+};
+
+layout(buffer_reference_align = 16, std430, buffer_reference) writeonly buffer OutputRootParameters
+{
+	RootParameterBlock blocks[];
+};
+
+layout(buffer_reference_align = 4, std430, buffer_reference) readonly buffer Template
+{
+	uint input_root_parameters[64];
+	int source_word_offsets[64];
+};
+
+layout(push_constant) uniform Registers
+{
+	Indirect indirect;
+	uvec2 count_va;
+	IndirectOutput out_dispatch;
+	OutputRootParameters out_root_parameters;
+	Template in_template;
+	uint stride_words;
+	uint dispatch_offset_words;
+};
+
+void main()
+{
+	bool active_dispatch;
+	if (any(notEqual(count_va, uvec2(0))))
+		active_dispatch = gl_WorkGroupID.x < IndirectCount(count_va).value;
+	else
+		active_dispatch = true;
+
+	uint input_offset_base = gl_WorkGroupID.x * stride_words;
+
+	uint new_root_parameter;
+	int source_offset = in_template.source_word_offsets[gl_LocalInvocationIndex];
+	if (source_offset >= 0)
+		new_root_parameter = indirect.values[input_offset_base + source_offset];
+	else
+		new_root_parameter = in_template.input_root_parameters[gl_LocalInvocationIndex];
+
+	out_root_parameters.blocks[gl_WorkGroupID.x].values[gl_LocalInvocationIndex] = new_root_parameter;
+
+	// TODO: We can change this constant to suit our needs if we need multi-indirect any kind of dispatch really.
+	// Spec constant is good for that scenario.
+
+	const uint DISPATCH_NUM_WORDS = 3;
+	if (gl_LocalInvocationIndex < DISPATCH_NUM_WORDS)
+	{
+		uint input_value;
+		if (active_dispatch)
+			input_value = indirect.values[input_offset_base + dispatch_offset_words + gl_LocalInvocationIndex];
+		else
+			input_value = 0u;
+
+		out_dispatch.values[DISPATCH_NUM_WORDS * gl_WorkGroupID.x + gl_LocalInvocationIndex] = input_value;
+	}
+}
+
diff --git a/libs/vkd3d/state.c b/libs/vkd3d/state.c
index ce7f54e..bca25c2 100644
--- a/libs/vkd3d/state.c
+++ b/libs/vkd3d/state.c
@@ -1487,6 +1487,8 @@ static HRESULT d3d12_root_signature_init_global(struct d3d12_root_signature *roo
     /* If we cannot contain the push constants, fall back to push UBO everywhere. */
     if (push_constant_range.size > vk_device_properties->limits.maxPushConstantsSize)
         d3d12_root_signature_add_common_flags(root_signature, VKD3D_ROOT_SIGNATURE_USE_PUSH_CONSTANT_UNIFORM_BLOCK);
+    else if (push_constant_range.size && (device->bindless_state.flags & VKD3D_FORCE_COMPUTE_ROOT_PARAMETERS_PUSH_UBO))
+        root_signature->compute.flags |= VKD3D_ROOT_SIGNATURE_USE_PUSH_CONSTANT_UNIFORM_BLOCK;
 
     d3d12_root_signature_init_extra_bindings(root_signature, &info);
 
@@ -7315,8 +7317,25 @@ static uint32_t vkd3d_bindless_state_get_bindless_flags(struct d3d12_device *dev
         flags |= VKD3D_HOIST_STATIC_TABLE_CBV;
     }
 
-    if (device->device_info.device_generated_commands_features.deviceGeneratedCommands)
+    if ((vkd3d_config_flags & VKD3D_CONFIG_FLAG_REQUIRES_COMPUTE_INDIRECT_TEMPLATES) &&
+            !device->device_info.device_generated_commands_compute_features_nv.deviceGeneratedCompute &&
+            !device->device_info.device_generated_commands_features_ext.deviceGeneratedCommands)
+    {
+        INFO("Forcing push UBO path for compute root parameters.\n");
+        flags |= VKD3D_FORCE_COMPUTE_ROOT_PARAMETERS_PUSH_UBO;
+    }
+
+    if (device->device_info.device_generated_commands_features_ext.deviceGeneratedCommands)
+    {
         INFO("Enabling fast paths for advanced ExecuteIndirect() graphics and compute (EXT_dgc).\n");
+    }
+    else
+    {
+        if (device->device_info.device_generated_commands_compute_features_nv.deviceGeneratedCompute)
+            INFO("Enabling fast paths for advanced ExecuteIndirect() compute (NV_dgc).\n");
+        if (device->device_info.device_generated_commands_features_nv.deviceGeneratedCommands)
+            INFO("Enabling fast paths for advanced ExecuteIndirect() graphics (NV_dgc).\n");
+    }
 
     if (vkd3d_bindless_supports_mutable_type(device, flags))
     {
diff --git a/libs/vkd3d/vkd3d_private.h b/libs/vkd3d/vkd3d_private.h
index 63154f9..24158d7 100644
--- a/libs/vkd3d/vkd3d_private.h
+++ b/libs/vkd3d/vkd3d_private.h
@@ -204,8 +204,10 @@ struct vkd3d_vulkan_info
     bool NVX_image_view_handle;
     bool NV_compute_shader_derivatives;
     bool NV_device_diagnostic_checkpoints;
+    bool NV_device_generated_commands;
     bool NV_shader_subgroup_partitioned;
     bool NV_memory_decompression;
+    bool NV_device_generated_commands_compute;
     bool NV_low_latency2;
     bool NV_raw_access_chains;
     bool NV_cooperative_matrix2;
@@ -3797,8 +3799,10 @@ struct d3d12_command_signature
             VkBuffer buffer;
             VkDeviceAddress buffer_va;
             struct vkd3d_device_memory_allocation memory;
-            VkIndirectCommandsLayoutEXT layout_implicit;
-            VkIndirectCommandsLayoutEXT layout_preprocess;
+            VkIndirectCommandsLayoutNV layout_implicit_nv;
+            VkIndirectCommandsLayoutNV layout_preprocess_nv;
+            VkIndirectCommandsLayoutEXT layout_implicit_ext;
+            VkIndirectCommandsLayoutEXT layout_preprocess_ext;
             uint32_t stride;
             struct vkd3d_execute_indirect_info pipeline;
         } dgc;
@@ -3808,6 +3812,7 @@ struct d3d12_command_signature
             uint32_t dispatch_offset_words;
         } compute;
     } state_template;
+    bool requires_state_template_dgc;
     bool requires_state_template;
     enum vkd3d_pipeline_type pipeline_type;
 
@@ -4270,7 +4275,8 @@ enum vkd3d_bindless_flags
     VKD3D_BINDLESS_MUTABLE_TYPE_RAW_SSBO            = (1u << 8),
     VKD3D_BINDLESS_MUTABLE_EMBEDDED                 = (1u << 9),
     VKD3D_BINDLESS_MUTABLE_EMBEDDED_PACKED_METADATA = (1u << 10),
-    VKD3D_BINDLESS_MUTABLE_TYPE_SPLIT_RAW_TYPED     = (1u << 11),
+    VKD3D_FORCE_COMPUTE_ROOT_PARAMETERS_PUSH_UBO    = (1u << 11),
+    VKD3D_BINDLESS_MUTABLE_TYPE_SPLIT_RAW_TYPED     = (1u << 12),
 };
 
 #define VKD3D_BINDLESS_SET_MAX_EXTRA_BINDINGS 8
@@ -4748,10 +4754,23 @@ struct vkd3d_multi_dispatch_indirect_args
     uint32_t max_commands;
 };
 
+struct vkd3d_multi_dispatch_indirect_state_args
+{
+    VkDeviceAddress indirect_va;
+    VkDeviceAddress count_va;
+    VkDeviceAddress dispatch_va;
+    VkDeviceAddress root_parameters_va;
+    VkDeviceAddress root_parameter_template_va;
+    uint32_t stride_words;
+    uint32_t dispatch_offset_words;
+};
+
 struct vkd3d_multi_dispatch_indirect_ops
 {
     VkPipelineLayout vk_multi_dispatch_indirect_layout;
+    VkPipelineLayout vk_multi_dispatch_indirect_state_layout;
     VkPipeline vk_multi_dispatch_indirect_pipeline;
+    VkPipeline vk_multi_dispatch_indirect_state_pipeline;
 };
 
 struct vkd3d_execute_indirect_args
@@ -4959,6 +4978,8 @@ void vkd3d_meta_get_predicate_pipeline(struct vkd3d_meta_ops *meta_ops,
 
 void vkd3d_meta_get_multi_dispatch_indirect_pipeline(struct vkd3d_meta_ops *meta_ops,
         struct vkd3d_multi_dispatch_indirect_info *info);
+void vkd3d_meta_get_multi_dispatch_indirect_state_pipeline(struct vkd3d_meta_ops *meta_ops,
+        struct vkd3d_multi_dispatch_indirect_info *info);
 
 static inline uint32_t vkd3d_meta_get_multi_dispatch_indirect_workgroup_size(void)
 {
@@ -5027,7 +5048,8 @@ struct vkd3d_physical_device_info
     VkPhysicalDeviceAccelerationStructurePropertiesKHR acceleration_structure_properties;
     VkPhysicalDeviceFragmentShadingRatePropertiesKHR fragment_shading_rate_properties;
     VkPhysicalDeviceConservativeRasterizationPropertiesEXT conservative_rasterization_properties;
-    VkPhysicalDeviceDeviceGeneratedCommandsPropertiesEXT device_generated_commands_properties;
+    VkPhysicalDeviceDeviceGeneratedCommandsPropertiesNV device_generated_commands_properties_nv;
+    VkPhysicalDeviceDeviceGeneratedCommandsPropertiesEXT device_generated_commands_properties_ext;
     VkPhysicalDeviceMeshShaderPropertiesEXT mesh_shader_properties;
     VkPhysicalDeviceShaderModuleIdentifierPropertiesEXT shader_module_identifier_properties;
     VkPhysicalDeviceDescriptorBufferPropertiesEXT descriptor_buffer_properties;
@@ -5067,7 +5089,8 @@ struct vkd3d_physical_device_info
     VkPhysicalDeviceImageViewMinLodFeaturesEXT image_view_min_lod_features;
     VkPhysicalDeviceCoherentMemoryFeaturesAMD device_coherent_memory_features_amd;
     VkPhysicalDeviceRayTracingMaintenance1FeaturesKHR ray_tracing_maintenance1_features;
-    VkPhysicalDeviceDeviceGeneratedCommandsFeaturesEXT device_generated_commands_features;
+    VkPhysicalDeviceDeviceGeneratedCommandsFeaturesNV device_generated_commands_features_nv;
+    VkPhysicalDeviceDeviceGeneratedCommandsFeaturesEXT device_generated_commands_features_ext;
     VkPhysicalDeviceMeshShaderFeaturesEXT mesh_shader_features;
     VkPhysicalDeviceShaderModuleIdentifierFeaturesEXT shader_module_identifier_features;
     VkPhysicalDevicePresentIdFeaturesKHR present_id_features;
@@ -5081,6 +5104,7 @@ struct vkd3d_physical_device_info
     VkPhysicalDevicePageableDeviceLocalMemoryFeaturesEXT pageable_device_memory_features;
     VkPhysicalDeviceDynamicRenderingUnusedAttachmentsFeaturesEXT dynamic_rendering_unused_attachments_features;
     VkPhysicalDeviceMemoryDecompressionFeaturesNV memory_decompression_features;
+    VkPhysicalDeviceDeviceGeneratedCommandsComputeFeaturesNV device_generated_commands_compute_features_nv;
     VkPhysicalDeviceMaintenance5FeaturesKHR maintenance_5_features;
     VkPhysicalDeviceMaintenance6FeaturesKHR maintenance_6_features;
     VkPhysicalDeviceMaintenance7FeaturesKHR maintenance_7_features;
diff --git a/libs/vkd3d/vkd3d_shaders.h b/libs/vkd3d/vkd3d_shaders.h
index ded6b46..79a462d 100644
--- a/libs/vkd3d/vkd3d_shaders.h
+++ b/libs/vkd3d/vkd3d_shaders.h
@@ -50,6 +50,7 @@ enum vkd3d_meta_copy_mode
 #include <cs_execute_indirect_patch.h>
 #include <cs_execute_indirect_patch_debug_ring.h>
 #include <cs_execute_indirect_multi_dispatch.h>
+#include <cs_execute_indirect_multi_dispatch_state.h>
 #include <vs_fullscreen_layer.h>
 #include <vs_fullscreen.h>
 #include <gs_fullscreen.h>
